// @cloudinary/html/index.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.CloudinaryHtml = {}));
}(this, (function (exports) { 'use strict';

	var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

	/**
	 * Removes all key-value entries from the list cache.
	 *
	 * @private
	 * @name clear
	 * @memberOf ListCache
	 */

	function listCacheClear$1() {
	  this.__data__ = [];
	  this.size = 0;
	}

	var _listCacheClear = listCacheClear$1;

	/**
	 * Performs a
	 * [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
	 * comparison between two values to determine if they are equivalent.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to compare.
	 * @param {*} other The other value to compare.
	 * @returns {boolean} Returns `true` if the values are equivalent, else `false`.
	 * @example
	 *
	 * var object = { 'a': 1 };
	 * var other = { 'a': 1 };
	 *
	 * _.eq(object, object);
	 * // => true
	 *
	 * _.eq(object, other);
	 * // => false
	 *
	 * _.eq('a', 'a');
	 * // => true
	 *
	 * _.eq('a', Object('a'));
	 * // => false
	 *
	 * _.eq(NaN, NaN);
	 * // => true
	 */

	function eq$2(value, other) {
	  return value === other || (value !== value && other !== other);
	}

	var eq_1 = eq$2;

	var eq$1 = eq_1;

	/**
	 * Gets the index at which the `key` is found in `array` of key-value pairs.
	 *
	 * @private
	 * @param {Array} array The array to inspect.
	 * @param {*} key The key to search for.
	 * @returns {number} Returns the index of the matched value, else `-1`.
	 */
	function assocIndexOf$4(array, key) {
	  var length = array.length;
	  while (length--) {
	    if (eq$1(array[length][0], key)) {
	      return length;
	    }
	  }
	  return -1;
	}

	var _assocIndexOf = assocIndexOf$4;

	var assocIndexOf$3 = _assocIndexOf;

	/** Used for built-in method references. */
	var arrayProto = Array.prototype;

	/** Built-in value references. */
	var splice = arrayProto.splice;

	/**
	 * Removes `key` and its value from the list cache.
	 *
	 * @private
	 * @name delete
	 * @memberOf ListCache
	 * @param {string} key The key of the value to remove.
	 * @returns {boolean} Returns `true` if the entry was removed, else `false`.
	 */
	function listCacheDelete$1(key) {
	  var data = this.__data__,
	      index = assocIndexOf$3(data, key);

	  if (index < 0) {
	    return false;
	  }
	  var lastIndex = data.length - 1;
	  if (index == lastIndex) {
	    data.pop();
	  } else {
	    splice.call(data, index, 1);
	  }
	  --this.size;
	  return true;
	}

	var _listCacheDelete = listCacheDelete$1;

	var assocIndexOf$2 = _assocIndexOf;

	/**
	 * Gets the list cache value for `key`.
	 *
	 * @private
	 * @name get
	 * @memberOf ListCache
	 * @param {string} key The key of the value to get.
	 * @returns {*} Returns the entry value.
	 */
	function listCacheGet$1(key) {
	  var data = this.__data__,
	      index = assocIndexOf$2(data, key);

	  return index < 0 ? undefined : data[index][1];
	}

	var _listCacheGet = listCacheGet$1;

	var assocIndexOf$1 = _assocIndexOf;

	/**
	 * Checks if a list cache value for `key` exists.
	 *
	 * @private
	 * @name has
	 * @memberOf ListCache
	 * @param {string} key The key of the entry to check.
	 * @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
	 */
	function listCacheHas$1(key) {
	  return assocIndexOf$1(this.__data__, key) > -1;
	}

	var _listCacheHas = listCacheHas$1;

	var assocIndexOf = _assocIndexOf;

	/**
	 * Sets the list cache `key` to `value`.
	 *
	 * @private
	 * @name set
	 * @memberOf ListCache
	 * @param {string} key The key of the value to set.
	 * @param {*} value The value to set.
	 * @returns {Object} Returns the list cache instance.
	 */
	function listCacheSet$1(key, value) {
	  var data = this.__data__,
	      index = assocIndexOf(data, key);

	  if (index < 0) {
	    ++this.size;
	    data.push([key, value]);
	  } else {
	    data[index][1] = value;
	  }
	  return this;
	}

	var _listCacheSet = listCacheSet$1;

	var listCacheClear = _listCacheClear,
	    listCacheDelete = _listCacheDelete,
	    listCacheGet = _listCacheGet,
	    listCacheHas = _listCacheHas,
	    listCacheSet = _listCacheSet;

	/**
	 * Creates an list cache object.
	 *
	 * @private
	 * @constructor
	 * @param {Array} [entries] The key-value pairs to cache.
	 */
	function ListCache$4(entries) {
	  var index = -1,
	      length = entries == null ? 0 : entries.length;

	  this.clear();
	  while (++index < length) {
	    var entry = entries[index];
	    this.set(entry[0], entry[1]);
	  }
	}

	// Add methods to `ListCache`.
	ListCache$4.prototype.clear = listCacheClear;
	ListCache$4.prototype['delete'] = listCacheDelete;
	ListCache$4.prototype.get = listCacheGet;
	ListCache$4.prototype.has = listCacheHas;
	ListCache$4.prototype.set = listCacheSet;

	var _ListCache = ListCache$4;

	var ListCache$3 = _ListCache;

	/**
	 * Removes all key-value entries from the stack.
	 *
	 * @private
	 * @name clear
	 * @memberOf Stack
	 */
	function stackClear$1() {
	  this.__data__ = new ListCache$3;
	  this.size = 0;
	}

	var _stackClear = stackClear$1;

	/**
	 * Removes `key` and its value from the stack.
	 *
	 * @private
	 * @name delete
	 * @memberOf Stack
	 * @param {string} key The key of the value to remove.
	 * @returns {boolean} Returns `true` if the entry was removed, else `false`.
	 */

	function stackDelete$1(key) {
	  var data = this.__data__,
	      result = data['delete'](key);

	  this.size = data.size;
	  return result;
	}

	var _stackDelete = stackDelete$1;

	/**
	 * Gets the stack value for `key`.
	 *
	 * @private
	 * @name get
	 * @memberOf Stack
	 * @param {string} key The key of the value to get.
	 * @returns {*} Returns the entry value.
	 */

	function stackGet$1(key) {
	  return this.__data__.get(key);
	}

	var _stackGet = stackGet$1;

	/**
	 * Checks if a stack value for `key` exists.
	 *
	 * @private
	 * @name has
	 * @memberOf Stack
	 * @param {string} key The key of the entry to check.
	 * @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
	 */

	function stackHas$1(key) {
	  return this.__data__.has(key);
	}

	var _stackHas = stackHas$1;

	/** Detect free variable `global` from Node.js. */

	var freeGlobal$2 = typeof commonjsGlobal == 'object' && commonjsGlobal && commonjsGlobal.Object === Object && commonjsGlobal;

	var _freeGlobal = freeGlobal$2;

	var freeGlobal$1 = _freeGlobal;

	/** Detect free variable `self`. */
	var freeSelf$1 = typeof self == 'object' && self && self.Object === Object && self;

	/** Used as a reference to the global object. */
	var root$9 = freeGlobal$1 || freeSelf$1 || Function('return this')();

	var _root = root$9;

	var root$8 = _root;

	/** Built-in value references. */
	var Symbol$4 = root$8.Symbol;

	var _Symbol = Symbol$4;

	var Symbol$3 = _Symbol;

	/** Used for built-in method references. */
	var objectProto$d = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$9 = objectProto$d.hasOwnProperty;

	/**
	 * Used to resolve the
	 * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
	 * of values.
	 */
	var nativeObjectToString$1 = objectProto$d.toString;

	/** Built-in value references. */
	var symToStringTag$1 = Symbol$3 ? Symbol$3.toStringTag : undefined;

	/**
	 * A specialized version of `baseGetTag` which ignores `Symbol.toStringTag` values.
	 *
	 * @private
	 * @param {*} value The value to query.
	 * @returns {string} Returns the raw `toStringTag`.
	 */
	function getRawTag$1(value) {
	  var isOwn = hasOwnProperty$9.call(value, symToStringTag$1),
	      tag = value[symToStringTag$1];

	  try {
	    value[symToStringTag$1] = undefined;
	    var unmasked = true;
	  } catch (e) {}

	  var result = nativeObjectToString$1.call(value);
	  if (unmasked) {
	    if (isOwn) {
	      value[symToStringTag$1] = tag;
	    } else {
	      delete value[symToStringTag$1];
	    }
	  }
	  return result;
	}

	var _getRawTag = getRawTag$1;

	/** Used for built-in method references. */

	var objectProto$c = Object.prototype;

	/**
	 * Used to resolve the
	 * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
	 * of values.
	 */
	var nativeObjectToString = objectProto$c.toString;

	/**
	 * Converts `value` to a string using `Object.prototype.toString`.
	 *
	 * @private
	 * @param {*} value The value to convert.
	 * @returns {string} Returns the converted string.
	 */
	function objectToString$2(value) {
	  return nativeObjectToString.call(value);
	}

	var _objectToString = objectToString$2;

	var Symbol$2 = _Symbol,
	    getRawTag = _getRawTag,
	    objectToString$1 = _objectToString;

	/** `Object#toString` result references. */
	var nullTag = '[object Null]',
	    undefinedTag = '[object Undefined]';

	/** Built-in value references. */
	var symToStringTag = Symbol$2 ? Symbol$2.toStringTag : undefined;

	/**
	 * The base implementation of `getTag` without fallbacks for buggy environments.
	 *
	 * @private
	 * @param {*} value The value to query.
	 * @returns {string} Returns the `toStringTag`.
	 */
	function baseGetTag$4(value) {
	  if (value == null) {
	    return value === undefined ? undefinedTag : nullTag;
	  }
	  return (symToStringTag && symToStringTag in Object(value))
	    ? getRawTag(value)
	    : objectToString$1(value);
	}

	var _baseGetTag = baseGetTag$4;

	/**
	 * Checks if `value` is the
	 * [language type](http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types)
	 * of `Object`. (e.g. arrays, functions, objects, regexes, `new Number(0)`, and `new String('')`)
	 *
	 * @static
	 * @memberOf _
	 * @since 0.1.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is an object, else `false`.
	 * @example
	 *
	 * _.isObject({});
	 * // => true
	 *
	 * _.isObject([1, 2, 3]);
	 * // => true
	 *
	 * _.isObject(_.noop);
	 * // => true
	 *
	 * _.isObject(null);
	 * // => false
	 */

	function isObject$6(value) {
	  var type = typeof value;
	  return value != null && (type == 'object' || type == 'function');
	}

	var isObject_1 = isObject$6;

	var baseGetTag$3 = _baseGetTag,
	    isObject$5 = isObject_1;

	/** `Object#toString` result references. */
	var asyncTag = '[object AsyncFunction]',
	    funcTag$2 = '[object Function]',
	    genTag$1 = '[object GeneratorFunction]',
	    proxyTag = '[object Proxy]';

	/**
	 * Checks if `value` is classified as a `Function` object.
	 *
	 * @static
	 * @memberOf _
	 * @since 0.1.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a function, else `false`.
	 * @example
	 *
	 * _.isFunction(_);
	 * // => true
	 *
	 * _.isFunction(/abc/);
	 * // => false
	 */
	function isFunction$2(value) {
	  if (!isObject$5(value)) {
	    return false;
	  }
	  // The use of `Object#toString` avoids issues with the `typeof` operator
	  // in Safari 9 which returns 'object' for typed arrays and other constructors.
	  var tag = baseGetTag$3(value);
	  return tag == funcTag$2 || tag == genTag$1 || tag == asyncTag || tag == proxyTag;
	}

	var isFunction_1 = isFunction$2;

	var root$7 = _root;

	/** Used to detect overreaching core-js shims. */
	var coreJsData$1 = root$7['__core-js_shared__'];

	var _coreJsData = coreJsData$1;

	var coreJsData = _coreJsData;

	/** Used to detect methods masquerading as native. */
	var maskSrcKey = (function() {
	  var uid = /[^.]+$/.exec(coreJsData && coreJsData.keys && coreJsData.keys.IE_PROTO || '');
	  return uid ? ('Symbol(src)_1.' + uid) : '';
	}());

	/**
	 * Checks if `func` has its source masked.
	 *
	 * @private
	 * @param {Function} func The function to check.
	 * @returns {boolean} Returns `true` if `func` is masked, else `false`.
	 */
	function isMasked$1(func) {
	  return !!maskSrcKey && (maskSrcKey in func);
	}

	var _isMasked = isMasked$1;

	/** Used for built-in method references. */

	var funcProto$1 = Function.prototype;

	/** Used to resolve the decompiled source of functions. */
	var funcToString$1 = funcProto$1.toString;

	/**
	 * Converts `func` to its source code.
	 *
	 * @private
	 * @param {Function} func The function to convert.
	 * @returns {string} Returns the source code.
	 */
	function toSource$2(func) {
	  if (func != null) {
	    try {
	      return funcToString$1.call(func);
	    } catch (e) {}
	    try {
	      return (func + '');
	    } catch (e) {}
	  }
	  return '';
	}

	var _toSource = toSource$2;

	var isFunction$1 = isFunction_1,
	    isMasked = _isMasked,
	    isObject$4 = isObject_1,
	    toSource$1 = _toSource;

	/**
	 * Used to match `RegExp`
	 * [syntax characters](http://ecma-international.org/ecma-262/7.0/#sec-patterns).
	 */
	var reRegExpChar = /[\\^$.*+?()[\]{}|]/g;

	/** Used to detect host constructors (Safari). */
	var reIsHostCtor = /^\[object .+?Constructor\]$/;

	/** Used for built-in method references. */
	var funcProto = Function.prototype,
	    objectProto$b = Object.prototype;

	/** Used to resolve the decompiled source of functions. */
	var funcToString = funcProto.toString;

	/** Used to check objects for own properties. */
	var hasOwnProperty$8 = objectProto$b.hasOwnProperty;

	/** Used to detect if a method is native. */
	var reIsNative = RegExp('^' +
	  funcToString.call(hasOwnProperty$8).replace(reRegExpChar, '\\$&')
	  .replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, '$1.*?') + '$'
	);

	/**
	 * The base implementation of `_.isNative` without bad shim checks.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a native function,
	 *  else `false`.
	 */
	function baseIsNative$1(value) {
	  if (!isObject$4(value) || isMasked(value)) {
	    return false;
	  }
	  var pattern = isFunction$1(value) ? reIsNative : reIsHostCtor;
	  return pattern.test(toSource$1(value));
	}

	var _baseIsNative = baseIsNative$1;

	/**
	 * Gets the value at `key` of `object`.
	 *
	 * @private
	 * @param {Object} [object] The object to query.
	 * @param {string} key The key of the property to get.
	 * @returns {*} Returns the property value.
	 */

	function getValue$1(object, key) {
	  return object == null ? undefined : object[key];
	}

	var _getValue = getValue$1;

	var baseIsNative = _baseIsNative,
	    getValue = _getValue;

	/**
	 * Gets the native function at `key` of `object`.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @param {string} key The key of the method to get.
	 * @returns {*} Returns the function if it's native, else `undefined`.
	 */
	function getNative$7(object, key) {
	  var value = getValue(object, key);
	  return baseIsNative(value) ? value : undefined;
	}

	var _getNative = getNative$7;

	var getNative$6 = _getNative,
	    root$6 = _root;

	/* Built-in method references that are verified to be native. */
	var Map$4 = getNative$6(root$6, 'Map');

	var _Map = Map$4;

	var getNative$5 = _getNative;

	/* Built-in method references that are verified to be native. */
	var nativeCreate$4 = getNative$5(Object, 'create');

	var _nativeCreate = nativeCreate$4;

	var nativeCreate$3 = _nativeCreate;

	/**
	 * Removes all key-value entries from the hash.
	 *
	 * @private
	 * @name clear
	 * @memberOf Hash
	 */
	function hashClear$1() {
	  this.__data__ = nativeCreate$3 ? nativeCreate$3(null) : {};
	  this.size = 0;
	}

	var _hashClear = hashClear$1;

	/**
	 * Removes `key` and its value from the hash.
	 *
	 * @private
	 * @name delete
	 * @memberOf Hash
	 * @param {Object} hash The hash to modify.
	 * @param {string} key The key of the value to remove.
	 * @returns {boolean} Returns `true` if the entry was removed, else `false`.
	 */

	function hashDelete$1(key) {
	  var result = this.has(key) && delete this.__data__[key];
	  this.size -= result ? 1 : 0;
	  return result;
	}

	var _hashDelete = hashDelete$1;

	var nativeCreate$2 = _nativeCreate;

	/** Used to stand-in for `undefined` hash values. */
	var HASH_UNDEFINED$1 = '__lodash_hash_undefined__';

	/** Used for built-in method references. */
	var objectProto$a = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$7 = objectProto$a.hasOwnProperty;

	/**
	 * Gets the hash value for `key`.
	 *
	 * @private
	 * @name get
	 * @memberOf Hash
	 * @param {string} key The key of the value to get.
	 * @returns {*} Returns the entry value.
	 */
	function hashGet$1(key) {
	  var data = this.__data__;
	  if (nativeCreate$2) {
	    var result = data[key];
	    return result === HASH_UNDEFINED$1 ? undefined : result;
	  }
	  return hasOwnProperty$7.call(data, key) ? data[key] : undefined;
	}

	var _hashGet = hashGet$1;

	var nativeCreate$1 = _nativeCreate;

	/** Used for built-in method references. */
	var objectProto$9 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$6 = objectProto$9.hasOwnProperty;

	/**
	 * Checks if a hash value for `key` exists.
	 *
	 * @private
	 * @name has
	 * @memberOf Hash
	 * @param {string} key The key of the entry to check.
	 * @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
	 */
	function hashHas$1(key) {
	  var data = this.__data__;
	  return nativeCreate$1 ? (data[key] !== undefined) : hasOwnProperty$6.call(data, key);
	}

	var _hashHas = hashHas$1;

	var nativeCreate = _nativeCreate;

	/** Used to stand-in for `undefined` hash values. */
	var HASH_UNDEFINED = '__lodash_hash_undefined__';

	/**
	 * Sets the hash `key` to `value`.
	 *
	 * @private
	 * @name set
	 * @memberOf Hash
	 * @param {string} key The key of the value to set.
	 * @param {*} value The value to set.
	 * @returns {Object} Returns the hash instance.
	 */
	function hashSet$1(key, value) {
	  var data = this.__data__;
	  this.size += this.has(key) ? 0 : 1;
	  data[key] = (nativeCreate && value === undefined) ? HASH_UNDEFINED : value;
	  return this;
	}

	var _hashSet = hashSet$1;

	var hashClear = _hashClear,
	    hashDelete = _hashDelete,
	    hashGet = _hashGet,
	    hashHas = _hashHas,
	    hashSet = _hashSet;

	/**
	 * Creates a hash object.
	 *
	 * @private
	 * @constructor
	 * @param {Array} [entries] The key-value pairs to cache.
	 */
	function Hash$1(entries) {
	  var index = -1,
	      length = entries == null ? 0 : entries.length;

	  this.clear();
	  while (++index < length) {
	    var entry = entries[index];
	    this.set(entry[0], entry[1]);
	  }
	}

	// Add methods to `Hash`.
	Hash$1.prototype.clear = hashClear;
	Hash$1.prototype['delete'] = hashDelete;
	Hash$1.prototype.get = hashGet;
	Hash$1.prototype.has = hashHas;
	Hash$1.prototype.set = hashSet;

	var _Hash = Hash$1;

	var Hash = _Hash,
	    ListCache$2 = _ListCache,
	    Map$3 = _Map;

	/**
	 * Removes all key-value entries from the map.
	 *
	 * @private
	 * @name clear
	 * @memberOf MapCache
	 */
	function mapCacheClear$1() {
	  this.size = 0;
	  this.__data__ = {
	    'hash': new Hash,
	    'map': new (Map$3 || ListCache$2),
	    'string': new Hash
	  };
	}

	var _mapCacheClear = mapCacheClear$1;

	/**
	 * Checks if `value` is suitable for use as unique object key.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is suitable, else `false`.
	 */

	function isKeyable$1(value) {
	  var type = typeof value;
	  return (type == 'string' || type == 'number' || type == 'symbol' || type == 'boolean')
	    ? (value !== '__proto__')
	    : (value === null);
	}

	var _isKeyable = isKeyable$1;

	var isKeyable = _isKeyable;

	/**
	 * Gets the data for `map`.
	 *
	 * @private
	 * @param {Object} map The map to query.
	 * @param {string} key The reference key.
	 * @returns {*} Returns the map data.
	 */
	function getMapData$4(map, key) {
	  var data = map.__data__;
	  return isKeyable(key)
	    ? data[typeof key == 'string' ? 'string' : 'hash']
	    : data.map;
	}

	var _getMapData = getMapData$4;

	var getMapData$3 = _getMapData;

	/**
	 * Removes `key` and its value from the map.
	 *
	 * @private
	 * @name delete
	 * @memberOf MapCache
	 * @param {string} key The key of the value to remove.
	 * @returns {boolean} Returns `true` if the entry was removed, else `false`.
	 */
	function mapCacheDelete$1(key) {
	  var result = getMapData$3(this, key)['delete'](key);
	  this.size -= result ? 1 : 0;
	  return result;
	}

	var _mapCacheDelete = mapCacheDelete$1;

	var getMapData$2 = _getMapData;

	/**
	 * Gets the map value for `key`.
	 *
	 * @private
	 * @name get
	 * @memberOf MapCache
	 * @param {string} key The key of the value to get.
	 * @returns {*} Returns the entry value.
	 */
	function mapCacheGet$1(key) {
	  return getMapData$2(this, key).get(key);
	}

	var _mapCacheGet = mapCacheGet$1;

	var getMapData$1 = _getMapData;

	/**
	 * Checks if a map value for `key` exists.
	 *
	 * @private
	 * @name has
	 * @memberOf MapCache
	 * @param {string} key The key of the entry to check.
	 * @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
	 */
	function mapCacheHas$1(key) {
	  return getMapData$1(this, key).has(key);
	}

	var _mapCacheHas = mapCacheHas$1;

	var getMapData = _getMapData;

	/**
	 * Sets the map `key` to `value`.
	 *
	 * @private
	 * @name set
	 * @memberOf MapCache
	 * @param {string} key The key of the value to set.
	 * @param {*} value The value to set.
	 * @returns {Object} Returns the map cache instance.
	 */
	function mapCacheSet$1(key, value) {
	  var data = getMapData(this, key),
	      size = data.size;

	  data.set(key, value);
	  this.size += data.size == size ? 0 : 1;
	  return this;
	}

	var _mapCacheSet = mapCacheSet$1;

	var mapCacheClear = _mapCacheClear,
	    mapCacheDelete = _mapCacheDelete,
	    mapCacheGet = _mapCacheGet,
	    mapCacheHas = _mapCacheHas,
	    mapCacheSet = _mapCacheSet;

	/**
	 * Creates a map cache object to store key-value pairs.
	 *
	 * @private
	 * @constructor
	 * @param {Array} [entries] The key-value pairs to cache.
	 */
	function MapCache$1(entries) {
	  var index = -1,
	      length = entries == null ? 0 : entries.length;

	  this.clear();
	  while (++index < length) {
	    var entry = entries[index];
	    this.set(entry[0], entry[1]);
	  }
	}

	// Add methods to `MapCache`.
	MapCache$1.prototype.clear = mapCacheClear;
	MapCache$1.prototype['delete'] = mapCacheDelete;
	MapCache$1.prototype.get = mapCacheGet;
	MapCache$1.prototype.has = mapCacheHas;
	MapCache$1.prototype.set = mapCacheSet;

	var _MapCache = MapCache$1;

	var ListCache$1 = _ListCache,
	    Map$2 = _Map,
	    MapCache = _MapCache;

	/** Used as the size to enable large array optimizations. */
	var LARGE_ARRAY_SIZE = 200;

	/**
	 * Sets the stack `key` to `value`.
	 *
	 * @private
	 * @name set
	 * @memberOf Stack
	 * @param {string} key The key of the value to set.
	 * @param {*} value The value to set.
	 * @returns {Object} Returns the stack cache instance.
	 */
	function stackSet$1(key, value) {
	  var data = this.__data__;
	  if (data instanceof ListCache$1) {
	    var pairs = data.__data__;
	    if (!Map$2 || (pairs.length < LARGE_ARRAY_SIZE - 1)) {
	      pairs.push([key, value]);
	      this.size = ++data.size;
	      return this;
	    }
	    data = this.__data__ = new MapCache(pairs);
	  }
	  data.set(key, value);
	  this.size = data.size;
	  return this;
	}

	var _stackSet = stackSet$1;

	var ListCache = _ListCache,
	    stackClear = _stackClear,
	    stackDelete = _stackDelete,
	    stackGet = _stackGet,
	    stackHas = _stackHas,
	    stackSet = _stackSet;

	/**
	 * Creates a stack cache object to store key-value pairs.
	 *
	 * @private
	 * @constructor
	 * @param {Array} [entries] The key-value pairs to cache.
	 */
	function Stack$1(entries) {
	  var data = this.__data__ = new ListCache(entries);
	  this.size = data.size;
	}

	// Add methods to `Stack`.
	Stack$1.prototype.clear = stackClear;
	Stack$1.prototype['delete'] = stackDelete;
	Stack$1.prototype.get = stackGet;
	Stack$1.prototype.has = stackHas;
	Stack$1.prototype.set = stackSet;

	var _Stack = Stack$1;

	/**
	 * A specialized version of `_.forEach` for arrays without support for
	 * iteratee shorthands.
	 *
	 * @private
	 * @param {Array} [array] The array to iterate over.
	 * @param {Function} iteratee The function invoked per iteration.
	 * @returns {Array} Returns `array`.
	 */

	function arrayEach$1(array, iteratee) {
	  var index = -1,
	      length = array == null ? 0 : array.length;

	  while (++index < length) {
	    if (iteratee(array[index], index, array) === false) {
	      break;
	    }
	  }
	  return array;
	}

	var _arrayEach = arrayEach$1;

	var getNative$4 = _getNative;

	var defineProperty$1 = (function() {
	  try {
	    var func = getNative$4(Object, 'defineProperty');
	    func({}, '', {});
	    return func;
	  } catch (e) {}
	}());

	var _defineProperty = defineProperty$1;

	var defineProperty = _defineProperty;

	/**
	 * The base implementation of `assignValue` and `assignMergeValue` without
	 * value checks.
	 *
	 * @private
	 * @param {Object} object The object to modify.
	 * @param {string} key The key of the property to assign.
	 * @param {*} value The value to assign.
	 */
	function baseAssignValue$2(object, key, value) {
	  if (key == '__proto__' && defineProperty) {
	    defineProperty(object, key, {
	      'configurable': true,
	      'enumerable': true,
	      'value': value,
	      'writable': true
	    });
	  } else {
	    object[key] = value;
	  }
	}

	var _baseAssignValue = baseAssignValue$2;

	var baseAssignValue$1 = _baseAssignValue,
	    eq = eq_1;

	/** Used for built-in method references. */
	var objectProto$8 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$5 = objectProto$8.hasOwnProperty;

	/**
	 * Assigns `value` to `key` of `object` if the existing value is not equivalent
	 * using [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
	 * for equality comparisons.
	 *
	 * @private
	 * @param {Object} object The object to modify.
	 * @param {string} key The key of the property to assign.
	 * @param {*} value The value to assign.
	 */
	function assignValue$2(object, key, value) {
	  var objValue = object[key];
	  if (!(hasOwnProperty$5.call(object, key) && eq(objValue, value)) ||
	      (value === undefined && !(key in object))) {
	    baseAssignValue$1(object, key, value);
	  }
	}

	var _assignValue = assignValue$2;

	var assignValue$1 = _assignValue,
	    baseAssignValue = _baseAssignValue;

	/**
	 * Copies properties of `source` to `object`.
	 *
	 * @private
	 * @param {Object} source The object to copy properties from.
	 * @param {Array} props The property identifiers to copy.
	 * @param {Object} [object={}] The object to copy properties to.
	 * @param {Function} [customizer] The function to customize copied values.
	 * @returns {Object} Returns `object`.
	 */
	function copyObject$4(source, props, object, customizer) {
	  var isNew = !object;
	  object || (object = {});

	  var index = -1,
	      length = props.length;

	  while (++index < length) {
	    var key = props[index];

	    var newValue = customizer
	      ? customizer(object[key], source[key], key, object, source)
	      : undefined;

	    if (newValue === undefined) {
	      newValue = source[key];
	    }
	    if (isNew) {
	      baseAssignValue(object, key, newValue);
	    } else {
	      assignValue$1(object, key, newValue);
	    }
	  }
	  return object;
	}

	var _copyObject = copyObject$4;

	/**
	 * The base implementation of `_.times` without support for iteratee shorthands
	 * or max array length checks.
	 *
	 * @private
	 * @param {number} n The number of times to invoke `iteratee`.
	 * @param {Function} iteratee The function invoked per iteration.
	 * @returns {Array} Returns the array of results.
	 */

	function baseTimes$1(n, iteratee) {
	  var index = -1,
	      result = Array(n);

	  while (++index < n) {
	    result[index] = iteratee(index);
	  }
	  return result;
	}

	var _baseTimes = baseTimes$1;

	/**
	 * Checks if `value` is object-like. A value is object-like if it's not `null`
	 * and has a `typeof` result of "object".
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is object-like, else `false`.
	 * @example
	 *
	 * _.isObjectLike({});
	 * // => true
	 *
	 * _.isObjectLike([1, 2, 3]);
	 * // => true
	 *
	 * _.isObjectLike(_.noop);
	 * // => false
	 *
	 * _.isObjectLike(null);
	 * // => false
	 */

	function isObjectLike$6(value) {
	  return value != null && typeof value == 'object';
	}

	var isObjectLike_1 = isObjectLike$6;

	var baseGetTag$2 = _baseGetTag,
	    isObjectLike$5 = isObjectLike_1;

	/** `Object#toString` result references. */
	var argsTag$2 = '[object Arguments]';

	/**
	 * The base implementation of `_.isArguments`.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is an `arguments` object,
	 */
	function baseIsArguments$1(value) {
	  return isObjectLike$5(value) && baseGetTag$2(value) == argsTag$2;
	}

	var _baseIsArguments = baseIsArguments$1;

	var baseIsArguments = _baseIsArguments,
	    isObjectLike$4 = isObjectLike_1;

	/** Used for built-in method references. */
	var objectProto$7 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$4 = objectProto$7.hasOwnProperty;

	/** Built-in value references. */
	var propertyIsEnumerable$1 = objectProto$7.propertyIsEnumerable;

	/**
	 * Checks if `value` is likely an `arguments` object.
	 *
	 * @static
	 * @memberOf _
	 * @since 0.1.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is an `arguments` object,
	 *  else `false`.
	 * @example
	 *
	 * _.isArguments(function() { return arguments; }());
	 * // => true
	 *
	 * _.isArguments([1, 2, 3]);
	 * // => false
	 */
	var isArguments$1 = baseIsArguments(function() { return arguments; }()) ? baseIsArguments : function(value) {
	  return isObjectLike$4(value) && hasOwnProperty$4.call(value, 'callee') &&
	    !propertyIsEnumerable$1.call(value, 'callee');
	};

	var isArguments_1 = isArguments$1;

	/**
	 * Checks if `value` is classified as an `Array` object.
	 *
	 * @static
	 * @memberOf _
	 * @since 0.1.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is an array, else `false`.
	 * @example
	 *
	 * _.isArray([1, 2, 3]);
	 * // => true
	 *
	 * _.isArray(document.body.children);
	 * // => false
	 *
	 * _.isArray('abc');
	 * // => false
	 *
	 * _.isArray(_.noop);
	 * // => false
	 */

	var isArray$3 = Array.isArray;

	var isArray_1 = isArray$3;

	var isBuffer$2 = {exports: {}};

	/**
	 * This method returns `false`.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.13.0
	 * @category Util
	 * @returns {boolean} Returns `false`.
	 * @example
	 *
	 * _.times(2, _.stubFalse);
	 * // => [false, false]
	 */

	function stubFalse() {
	  return false;
	}

	var stubFalse_1 = stubFalse;

	(function (module, exports) {
	var root = _root,
	    stubFalse = stubFalse_1;

	/** Detect free variable `exports`. */
	var freeExports = exports && !exports.nodeType && exports;

	/** Detect free variable `module`. */
	var freeModule = freeExports && 'object' == 'object' && module && !module.nodeType && module;

	/** Detect the popular CommonJS extension `module.exports`. */
	var moduleExports = freeModule && freeModule.exports === freeExports;

	/** Built-in value references. */
	var Buffer = moduleExports ? root.Buffer : undefined;

	/* Built-in method references for those with the same name as other `lodash` methods. */
	var nativeIsBuffer = Buffer ? Buffer.isBuffer : undefined;

	/**
	 * Checks if `value` is a buffer.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.3.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a buffer, else `false`.
	 * @example
	 *
	 * _.isBuffer(new Buffer(2));
	 * // => true
	 *
	 * _.isBuffer(new Uint8Array(2));
	 * // => false
	 */
	var isBuffer = nativeIsBuffer || stubFalse;

	module.exports = isBuffer;
	}(isBuffer$2, isBuffer$2.exports));

	/** Used as references for various `Number` constants. */

	var MAX_SAFE_INTEGER$1 = 9007199254740991;

	/** Used to detect unsigned integer values. */
	var reIsUint = /^(?:0|[1-9]\d*)$/;

	/**
	 * Checks if `value` is a valid array-like index.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @param {number} [length=MAX_SAFE_INTEGER] The upper bounds of a valid index.
	 * @returns {boolean} Returns `true` if `value` is a valid index, else `false`.
	 */
	function isIndex$1(value, length) {
	  var type = typeof value;
	  length = length == null ? MAX_SAFE_INTEGER$1 : length;

	  return !!length &&
	    (type == 'number' ||
	      (type != 'symbol' && reIsUint.test(value))) &&
	        (value > -1 && value % 1 == 0 && value < length);
	}

	var _isIndex = isIndex$1;

	/** Used as references for various `Number` constants. */

	var MAX_SAFE_INTEGER = 9007199254740991;

	/**
	 * Checks if `value` is a valid array-like length.
	 *
	 * **Note:** This method is loosely based on
	 * [`ToLength`](http://ecma-international.org/ecma-262/7.0/#sec-tolength).
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a valid length, else `false`.
	 * @example
	 *
	 * _.isLength(3);
	 * // => true
	 *
	 * _.isLength(Number.MIN_VALUE);
	 * // => false
	 *
	 * _.isLength(Infinity);
	 * // => false
	 *
	 * _.isLength('3');
	 * // => false
	 */
	function isLength$2(value) {
	  return typeof value == 'number' &&
	    value > -1 && value % 1 == 0 && value <= MAX_SAFE_INTEGER;
	}

	var isLength_1 = isLength$2;

	var baseGetTag$1 = _baseGetTag,
	    isLength$1 = isLength_1,
	    isObjectLike$3 = isObjectLike_1;

	/** `Object#toString` result references. */
	var argsTag$1 = '[object Arguments]',
	    arrayTag$1 = '[object Array]',
	    boolTag$2 = '[object Boolean]',
	    dateTag$2 = '[object Date]',
	    errorTag$1 = '[object Error]',
	    funcTag$1 = '[object Function]',
	    mapTag$4 = '[object Map]',
	    numberTag$2 = '[object Number]',
	    objectTag$2 = '[object Object]',
	    regexpTag$2 = '[object RegExp]',
	    setTag$4 = '[object Set]',
	    stringTag$2 = '[object String]',
	    weakMapTag$2 = '[object WeakMap]';

	var arrayBufferTag$2 = '[object ArrayBuffer]',
	    dataViewTag$3 = '[object DataView]',
	    float32Tag$2 = '[object Float32Array]',
	    float64Tag$2 = '[object Float64Array]',
	    int8Tag$2 = '[object Int8Array]',
	    int16Tag$2 = '[object Int16Array]',
	    int32Tag$2 = '[object Int32Array]',
	    uint8Tag$2 = '[object Uint8Array]',
	    uint8ClampedTag$2 = '[object Uint8ClampedArray]',
	    uint16Tag$2 = '[object Uint16Array]',
	    uint32Tag$2 = '[object Uint32Array]';

	/** Used to identify `toStringTag` values of typed arrays. */
	var typedArrayTags = {};
	typedArrayTags[float32Tag$2] = typedArrayTags[float64Tag$2] =
	typedArrayTags[int8Tag$2] = typedArrayTags[int16Tag$2] =
	typedArrayTags[int32Tag$2] = typedArrayTags[uint8Tag$2] =
	typedArrayTags[uint8ClampedTag$2] = typedArrayTags[uint16Tag$2] =
	typedArrayTags[uint32Tag$2] = true;
	typedArrayTags[argsTag$1] = typedArrayTags[arrayTag$1] =
	typedArrayTags[arrayBufferTag$2] = typedArrayTags[boolTag$2] =
	typedArrayTags[dataViewTag$3] = typedArrayTags[dateTag$2] =
	typedArrayTags[errorTag$1] = typedArrayTags[funcTag$1] =
	typedArrayTags[mapTag$4] = typedArrayTags[numberTag$2] =
	typedArrayTags[objectTag$2] = typedArrayTags[regexpTag$2] =
	typedArrayTags[setTag$4] = typedArrayTags[stringTag$2] =
	typedArrayTags[weakMapTag$2] = false;

	/**
	 * The base implementation of `_.isTypedArray` without Node.js optimizations.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a typed array, else `false`.
	 */
	function baseIsTypedArray$1(value) {
	  return isObjectLike$3(value) &&
	    isLength$1(value.length) && !!typedArrayTags[baseGetTag$1(value)];
	}

	var _baseIsTypedArray = baseIsTypedArray$1;

	/**
	 * The base implementation of `_.unary` without support for storing metadata.
	 *
	 * @private
	 * @param {Function} func The function to cap arguments for.
	 * @returns {Function} Returns the new capped function.
	 */

	function baseUnary$3(func) {
	  return function(value) {
	    return func(value);
	  };
	}

	var _baseUnary = baseUnary$3;

	var _nodeUtil = {exports: {}};

	(function (module, exports) {
	var freeGlobal = _freeGlobal;

	/** Detect free variable `exports`. */
	var freeExports = exports && !exports.nodeType && exports;

	/** Detect free variable `module`. */
	var freeModule = freeExports && 'object' == 'object' && module && !module.nodeType && module;

	/** Detect the popular CommonJS extension `module.exports`. */
	var moduleExports = freeModule && freeModule.exports === freeExports;

	/** Detect free variable `process` from Node.js. */
	var freeProcess = moduleExports && freeGlobal.process;

	/** Used to access faster Node.js helpers. */
	var nodeUtil = (function() {
	  try {
	    // Use `util.types` for Node.js 10+.
	    var types = freeModule && freeModule.require && freeModule.require('util').types;

	    if (types) {
	      return types;
	    }

	    // Legacy `process.binding('util')` for Node.js < 10.
	    return freeProcess && freeProcess.binding && freeProcess.binding('util');
	  } catch (e) {}
	}());

	module.exports = nodeUtil;
	}(_nodeUtil, _nodeUtil.exports));

	var baseIsTypedArray = _baseIsTypedArray,
	    baseUnary$2 = _baseUnary,
	    nodeUtil$2 = _nodeUtil.exports;

	/* Node.js helper references. */
	var nodeIsTypedArray = nodeUtil$2 && nodeUtil$2.isTypedArray;

	/**
	 * Checks if `value` is classified as a typed array.
	 *
	 * @static
	 * @memberOf _
	 * @since 3.0.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a typed array, else `false`.
	 * @example
	 *
	 * _.isTypedArray(new Uint8Array);
	 * // => true
	 *
	 * _.isTypedArray([]);
	 * // => false
	 */
	var isTypedArray$1 = nodeIsTypedArray ? baseUnary$2(nodeIsTypedArray) : baseIsTypedArray;

	var isTypedArray_1 = isTypedArray$1;

	var baseTimes = _baseTimes,
	    isArguments = isArguments_1,
	    isArray$2 = isArray_1,
	    isBuffer$1 = isBuffer$2.exports,
	    isIndex = _isIndex,
	    isTypedArray = isTypedArray_1;

	/** Used for built-in method references. */
	var objectProto$6 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$3 = objectProto$6.hasOwnProperty;

	/**
	 * Creates an array of the enumerable property names of the array-like `value`.
	 *
	 * @private
	 * @param {*} value The value to query.
	 * @param {boolean} inherited Specify returning inherited property names.
	 * @returns {Array} Returns the array of property names.
	 */
	function arrayLikeKeys$2(value, inherited) {
	  var isArr = isArray$2(value),
	      isArg = !isArr && isArguments(value),
	      isBuff = !isArr && !isArg && isBuffer$1(value),
	      isType = !isArr && !isArg && !isBuff && isTypedArray(value),
	      skipIndexes = isArr || isArg || isBuff || isType,
	      result = skipIndexes ? baseTimes(value.length, String) : [],
	      length = result.length;

	  for (var key in value) {
	    if ((inherited || hasOwnProperty$3.call(value, key)) &&
	        !(skipIndexes && (
	           // Safari 9 has enumerable `arguments.length` in strict mode.
	           key == 'length' ||
	           // Node.js 0.10 has enumerable non-index properties on buffers.
	           (isBuff && (key == 'offset' || key == 'parent')) ||
	           // PhantomJS 2 has enumerable non-index properties on typed arrays.
	           (isType && (key == 'buffer' || key == 'byteLength' || key == 'byteOffset')) ||
	           // Skip index properties.
	           isIndex(key, length)
	        ))) {
	      result.push(key);
	    }
	  }
	  return result;
	}

	var _arrayLikeKeys = arrayLikeKeys$2;

	/** Used for built-in method references. */

	var objectProto$5 = Object.prototype;

	/**
	 * Checks if `value` is likely a prototype object.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a prototype, else `false`.
	 */
	function isPrototype$3(value) {
	  var Ctor = value && value.constructor,
	      proto = (typeof Ctor == 'function' && Ctor.prototype) || objectProto$5;

	  return value === proto;
	}

	var _isPrototype = isPrototype$3;

	/**
	 * Creates a unary function that invokes `func` with its argument transformed.
	 *
	 * @private
	 * @param {Function} func The function to wrap.
	 * @param {Function} transform The argument transform.
	 * @returns {Function} Returns the new function.
	 */

	function overArg$2(func, transform) {
	  return function(arg) {
	    return func(transform(arg));
	  };
	}

	var _overArg = overArg$2;

	var overArg$1 = _overArg;

	/* Built-in method references for those with the same name as other `lodash` methods. */
	var nativeKeys$1 = overArg$1(Object.keys, Object);

	var _nativeKeys = nativeKeys$1;

	var isPrototype$2 = _isPrototype,
	    nativeKeys = _nativeKeys;

	/** Used for built-in method references. */
	var objectProto$4 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$2 = objectProto$4.hasOwnProperty;

	/**
	 * The base implementation of `_.keys` which doesn't treat sparse arrays as dense.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names.
	 */
	function baseKeys$1(object) {
	  if (!isPrototype$2(object)) {
	    return nativeKeys(object);
	  }
	  var result = [];
	  for (var key in Object(object)) {
	    if (hasOwnProperty$2.call(object, key) && key != 'constructor') {
	      result.push(key);
	    }
	  }
	  return result;
	}

	var _baseKeys = baseKeys$1;

	var isFunction = isFunction_1,
	    isLength = isLength_1;

	/**
	 * Checks if `value` is array-like. A value is considered array-like if it's
	 * not a function and has a `value.length` that's an integer greater than or
	 * equal to `0` and less than or equal to `Number.MAX_SAFE_INTEGER`.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is array-like, else `false`.
	 * @example
	 *
	 * _.isArrayLike([1, 2, 3]);
	 * // => true
	 *
	 * _.isArrayLike(document.body.children);
	 * // => true
	 *
	 * _.isArrayLike('abc');
	 * // => true
	 *
	 * _.isArrayLike(_.noop);
	 * // => false
	 */
	function isArrayLike$2(value) {
	  return value != null && isLength(value.length) && !isFunction(value);
	}

	var isArrayLike_1 = isArrayLike$2;

	var arrayLikeKeys$1 = _arrayLikeKeys,
	    baseKeys = _baseKeys,
	    isArrayLike$1 = isArrayLike_1;

	/**
	 * Creates an array of the own enumerable property names of `object`.
	 *
	 * **Note:** Non-object values are coerced to objects. See the
	 * [ES spec](http://ecma-international.org/ecma-262/7.0/#sec-object.keys)
	 * for more details.
	 *
	 * @static
	 * @since 0.1.0
	 * @memberOf _
	 * @category Object
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names.
	 * @example
	 *
	 * function Foo() {
	 *   this.a = 1;
	 *   this.b = 2;
	 * }
	 *
	 * Foo.prototype.c = 3;
	 *
	 * _.keys(new Foo);
	 * // => ['a', 'b'] (iteration order is not guaranteed)
	 *
	 * _.keys('hi');
	 * // => ['0', '1']
	 */
	function keys$3(object) {
	  return isArrayLike$1(object) ? arrayLikeKeys$1(object) : baseKeys(object);
	}

	var keys_1 = keys$3;

	var copyObject$3 = _copyObject,
	    keys$2 = keys_1;

	/**
	 * The base implementation of `_.assign` without support for multiple sources
	 * or `customizer` functions.
	 *
	 * @private
	 * @param {Object} object The destination object.
	 * @param {Object} source The source object.
	 * @returns {Object} Returns `object`.
	 */
	function baseAssign$1(object, source) {
	  return object && copyObject$3(source, keys$2(source), object);
	}

	var _baseAssign = baseAssign$1;

	/**
	 * This function is like
	 * [`Object.keys`](http://ecma-international.org/ecma-262/7.0/#sec-object.keys)
	 * except that it includes inherited enumerable properties.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names.
	 */

	function nativeKeysIn$1(object) {
	  var result = [];
	  if (object != null) {
	    for (var key in Object(object)) {
	      result.push(key);
	    }
	  }
	  return result;
	}

	var _nativeKeysIn = nativeKeysIn$1;

	var isObject$3 = isObject_1,
	    isPrototype$1 = _isPrototype,
	    nativeKeysIn = _nativeKeysIn;

	/** Used for built-in method references. */
	var objectProto$3 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty$1 = objectProto$3.hasOwnProperty;

	/**
	 * The base implementation of `_.keysIn` which doesn't treat sparse arrays as dense.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names.
	 */
	function baseKeysIn$1(object) {
	  if (!isObject$3(object)) {
	    return nativeKeysIn(object);
	  }
	  var isProto = isPrototype$1(object),
	      result = [];

	  for (var key in object) {
	    if (!(key == 'constructor' && (isProto || !hasOwnProperty$1.call(object, key)))) {
	      result.push(key);
	    }
	  }
	  return result;
	}

	var _baseKeysIn = baseKeysIn$1;

	var arrayLikeKeys = _arrayLikeKeys,
	    baseKeysIn = _baseKeysIn,
	    isArrayLike = isArrayLike_1;

	/**
	 * Creates an array of the own and inherited enumerable property names of `object`.
	 *
	 * **Note:** Non-object values are coerced to objects.
	 *
	 * @static
	 * @memberOf _
	 * @since 3.0.0
	 * @category Object
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names.
	 * @example
	 *
	 * function Foo() {
	 *   this.a = 1;
	 *   this.b = 2;
	 * }
	 *
	 * Foo.prototype.c = 3;
	 *
	 * _.keysIn(new Foo);
	 * // => ['a', 'b', 'c'] (iteration order is not guaranteed)
	 */
	function keysIn$3(object) {
	  return isArrayLike(object) ? arrayLikeKeys(object, true) : baseKeysIn(object);
	}

	var keysIn_1 = keysIn$3;

	var copyObject$2 = _copyObject,
	    keysIn$2 = keysIn_1;

	/**
	 * The base implementation of `_.assignIn` without support for multiple sources
	 * or `customizer` functions.
	 *
	 * @private
	 * @param {Object} object The destination object.
	 * @param {Object} source The source object.
	 * @returns {Object} Returns `object`.
	 */
	function baseAssignIn$1(object, source) {
	  return object && copyObject$2(source, keysIn$2(source), object);
	}

	var _baseAssignIn = baseAssignIn$1;

	var _cloneBuffer = {exports: {}};

	(function (module, exports) {
	var root = _root;

	/** Detect free variable `exports`. */
	var freeExports = exports && !exports.nodeType && exports;

	/** Detect free variable `module`. */
	var freeModule = freeExports && 'object' == 'object' && module && !module.nodeType && module;

	/** Detect the popular CommonJS extension `module.exports`. */
	var moduleExports = freeModule && freeModule.exports === freeExports;

	/** Built-in value references. */
	var Buffer = moduleExports ? root.Buffer : undefined,
	    allocUnsafe = Buffer ? Buffer.allocUnsafe : undefined;

	/**
	 * Creates a clone of  `buffer`.
	 *
	 * @private
	 * @param {Buffer} buffer The buffer to clone.
	 * @param {boolean} [isDeep] Specify a deep clone.
	 * @returns {Buffer} Returns the cloned buffer.
	 */
	function cloneBuffer(buffer, isDeep) {
	  if (isDeep) {
	    return buffer.slice();
	  }
	  var length = buffer.length,
	      result = allocUnsafe ? allocUnsafe(length) : new buffer.constructor(length);

	  buffer.copy(result);
	  return result;
	}

	module.exports = cloneBuffer;
	}(_cloneBuffer, _cloneBuffer.exports));

	/**
	 * Copies the values of `source` to `array`.
	 *
	 * @private
	 * @param {Array} source The array to copy values from.
	 * @param {Array} [array=[]] The array to copy values to.
	 * @returns {Array} Returns `array`.
	 */

	function copyArray$1(source, array) {
	  var index = -1,
	      length = source.length;

	  array || (array = Array(length));
	  while (++index < length) {
	    array[index] = source[index];
	  }
	  return array;
	}

	var _copyArray = copyArray$1;

	/**
	 * A specialized version of `_.filter` for arrays without support for
	 * iteratee shorthands.
	 *
	 * @private
	 * @param {Array} [array] The array to iterate over.
	 * @param {Function} predicate The function invoked per iteration.
	 * @returns {Array} Returns the new filtered array.
	 */

	function arrayFilter$1(array, predicate) {
	  var index = -1,
	      length = array == null ? 0 : array.length,
	      resIndex = 0,
	      result = [];

	  while (++index < length) {
	    var value = array[index];
	    if (predicate(value, index, array)) {
	      result[resIndex++] = value;
	    }
	  }
	  return result;
	}

	var _arrayFilter = arrayFilter$1;

	/**
	 * This method returns a new empty array.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.13.0
	 * @category Util
	 * @returns {Array} Returns the new empty array.
	 * @example
	 *
	 * var arrays = _.times(2, _.stubArray);
	 *
	 * console.log(arrays);
	 * // => [[], []]
	 *
	 * console.log(arrays[0] === arrays[1]);
	 * // => false
	 */

	function stubArray$2() {
	  return [];
	}

	var stubArray_1 = stubArray$2;

	var arrayFilter = _arrayFilter,
	    stubArray$1 = stubArray_1;

	/** Used for built-in method references. */
	var objectProto$2 = Object.prototype;

	/** Built-in value references. */
	var propertyIsEnumerable = objectProto$2.propertyIsEnumerable;

	/* Built-in method references for those with the same name as other `lodash` methods. */
	var nativeGetSymbols$1 = Object.getOwnPropertySymbols;

	/**
	 * Creates an array of the own enumerable symbols of `object`.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of symbols.
	 */
	var getSymbols$3 = !nativeGetSymbols$1 ? stubArray$1 : function(object) {
	  if (object == null) {
	    return [];
	  }
	  object = Object(object);
	  return arrayFilter(nativeGetSymbols$1(object), function(symbol) {
	    return propertyIsEnumerable.call(object, symbol);
	  });
	};

	var _getSymbols = getSymbols$3;

	var copyObject$1 = _copyObject,
	    getSymbols$2 = _getSymbols;

	/**
	 * Copies own symbols of `source` to `object`.
	 *
	 * @private
	 * @param {Object} source The object to copy symbols from.
	 * @param {Object} [object={}] The object to copy symbols to.
	 * @returns {Object} Returns `object`.
	 */
	function copySymbols$1(source, object) {
	  return copyObject$1(source, getSymbols$2(source), object);
	}

	var _copySymbols = copySymbols$1;

	/**
	 * Appends the elements of `values` to `array`.
	 *
	 * @private
	 * @param {Array} array The array to modify.
	 * @param {Array} values The values to append.
	 * @returns {Array} Returns `array`.
	 */

	function arrayPush$2(array, values) {
	  var index = -1,
	      length = values.length,
	      offset = array.length;

	  while (++index < length) {
	    array[offset + index] = values[index];
	  }
	  return array;
	}

	var _arrayPush = arrayPush$2;

	var overArg = _overArg;

	/** Built-in value references. */
	var getPrototype$2 = overArg(Object.getPrototypeOf, Object);

	var _getPrototype = getPrototype$2;

	var arrayPush$1 = _arrayPush,
	    getPrototype$1 = _getPrototype,
	    getSymbols$1 = _getSymbols,
	    stubArray = stubArray_1;

	/* Built-in method references for those with the same name as other `lodash` methods. */
	var nativeGetSymbols = Object.getOwnPropertySymbols;

	/**
	 * Creates an array of the own and inherited enumerable symbols of `object`.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of symbols.
	 */
	var getSymbolsIn$2 = !nativeGetSymbols ? stubArray : function(object) {
	  var result = [];
	  while (object) {
	    arrayPush$1(result, getSymbols$1(object));
	    object = getPrototype$1(object);
	  }
	  return result;
	};

	var _getSymbolsIn = getSymbolsIn$2;

	var copyObject = _copyObject,
	    getSymbolsIn$1 = _getSymbolsIn;

	/**
	 * Copies own and inherited symbols of `source` to `object`.
	 *
	 * @private
	 * @param {Object} source The object to copy symbols from.
	 * @param {Object} [object={}] The object to copy symbols to.
	 * @returns {Object} Returns `object`.
	 */
	function copySymbolsIn$1(source, object) {
	  return copyObject(source, getSymbolsIn$1(source), object);
	}

	var _copySymbolsIn = copySymbolsIn$1;

	var arrayPush = _arrayPush,
	    isArray$1 = isArray_1;

	/**
	 * The base implementation of `getAllKeys` and `getAllKeysIn` which uses
	 * `keysFunc` and `symbolsFunc` to get the enumerable property names and
	 * symbols of `object`.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @param {Function} keysFunc The function to get the keys of `object`.
	 * @param {Function} symbolsFunc The function to get the symbols of `object`.
	 * @returns {Array} Returns the array of property names and symbols.
	 */
	function baseGetAllKeys$2(object, keysFunc, symbolsFunc) {
	  var result = keysFunc(object);
	  return isArray$1(object) ? result : arrayPush(result, symbolsFunc(object));
	}

	var _baseGetAllKeys = baseGetAllKeys$2;

	var baseGetAllKeys$1 = _baseGetAllKeys,
	    getSymbols = _getSymbols,
	    keys$1 = keys_1;

	/**
	 * Creates an array of own enumerable property names and symbols of `object`.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names and symbols.
	 */
	function getAllKeys$1(object) {
	  return baseGetAllKeys$1(object, keys$1, getSymbols);
	}

	var _getAllKeys = getAllKeys$1;

	var baseGetAllKeys = _baseGetAllKeys,
	    getSymbolsIn = _getSymbolsIn,
	    keysIn$1 = keysIn_1;

	/**
	 * Creates an array of own and inherited enumerable property names and
	 * symbols of `object`.
	 *
	 * @private
	 * @param {Object} object The object to query.
	 * @returns {Array} Returns the array of property names and symbols.
	 */
	function getAllKeysIn$1(object) {
	  return baseGetAllKeys(object, keysIn$1, getSymbolsIn);
	}

	var _getAllKeysIn = getAllKeysIn$1;

	var getNative$3 = _getNative,
	    root$5 = _root;

	/* Built-in method references that are verified to be native. */
	var DataView$1 = getNative$3(root$5, 'DataView');

	var _DataView = DataView$1;

	var getNative$2 = _getNative,
	    root$4 = _root;

	/* Built-in method references that are verified to be native. */
	var Promise$2 = getNative$2(root$4, 'Promise');

	var _Promise = Promise$2;

	var getNative$1 = _getNative,
	    root$3 = _root;

	/* Built-in method references that are verified to be native. */
	var Set$1 = getNative$1(root$3, 'Set');

	var _Set = Set$1;

	var getNative = _getNative,
	    root$2 = _root;

	/* Built-in method references that are verified to be native. */
	var WeakMap$1 = getNative(root$2, 'WeakMap');

	var _WeakMap = WeakMap$1;

	var DataView = _DataView,
	    Map$1 = _Map,
	    Promise$1 = _Promise,
	    Set = _Set,
	    WeakMap = _WeakMap,
	    baseGetTag = _baseGetTag,
	    toSource = _toSource;

	/** `Object#toString` result references. */
	var mapTag$3 = '[object Map]',
	    objectTag$1 = '[object Object]',
	    promiseTag = '[object Promise]',
	    setTag$3 = '[object Set]',
	    weakMapTag$1 = '[object WeakMap]';

	var dataViewTag$2 = '[object DataView]';

	/** Used to detect maps, sets, and weakmaps. */
	var dataViewCtorString = toSource(DataView),
	    mapCtorString = toSource(Map$1),
	    promiseCtorString = toSource(Promise$1),
	    setCtorString = toSource(Set),
	    weakMapCtorString = toSource(WeakMap);

	/**
	 * Gets the `toStringTag` of `value`.
	 *
	 * @private
	 * @param {*} value The value to query.
	 * @returns {string} Returns the `toStringTag`.
	 */
	var getTag$3 = baseGetTag;

	// Fallback for data views, maps, sets, and weak maps in IE 11 and promises in Node.js < 6.
	if ((DataView && getTag$3(new DataView(new ArrayBuffer(1))) != dataViewTag$2) ||
	    (Map$1 && getTag$3(new Map$1) != mapTag$3) ||
	    (Promise$1 && getTag$3(Promise$1.resolve()) != promiseTag) ||
	    (Set && getTag$3(new Set) != setTag$3) ||
	    (WeakMap && getTag$3(new WeakMap) != weakMapTag$1)) {
	  getTag$3 = function(value) {
	    var result = baseGetTag(value),
	        Ctor = result == objectTag$1 ? value.constructor : undefined,
	        ctorString = Ctor ? toSource(Ctor) : '';

	    if (ctorString) {
	      switch (ctorString) {
	        case dataViewCtorString: return dataViewTag$2;
	        case mapCtorString: return mapTag$3;
	        case promiseCtorString: return promiseTag;
	        case setCtorString: return setTag$3;
	        case weakMapCtorString: return weakMapTag$1;
	      }
	    }
	    return result;
	  };
	}

	var _getTag = getTag$3;

	/** Used for built-in method references. */

	var objectProto$1 = Object.prototype;

	/** Used to check objects for own properties. */
	var hasOwnProperty = objectProto$1.hasOwnProperty;

	/**
	 * Initializes an array clone.
	 *
	 * @private
	 * @param {Array} array The array to clone.
	 * @returns {Array} Returns the initialized clone.
	 */
	function initCloneArray$1(array) {
	  var length = array.length,
	      result = new array.constructor(length);

	  // Add properties assigned by `RegExp#exec`.
	  if (length && typeof array[0] == 'string' && hasOwnProperty.call(array, 'index')) {
	    result.index = array.index;
	    result.input = array.input;
	  }
	  return result;
	}

	var _initCloneArray = initCloneArray$1;

	var root$1 = _root;

	/** Built-in value references. */
	var Uint8Array$1 = root$1.Uint8Array;

	var _Uint8Array = Uint8Array$1;

	var Uint8Array = _Uint8Array;

	/**
	 * Creates a clone of `arrayBuffer`.
	 *
	 * @private
	 * @param {ArrayBuffer} arrayBuffer The array buffer to clone.
	 * @returns {ArrayBuffer} Returns the cloned array buffer.
	 */
	function cloneArrayBuffer$3(arrayBuffer) {
	  var result = new arrayBuffer.constructor(arrayBuffer.byteLength);
	  new Uint8Array(result).set(new Uint8Array(arrayBuffer));
	  return result;
	}

	var _cloneArrayBuffer = cloneArrayBuffer$3;

	var cloneArrayBuffer$2 = _cloneArrayBuffer;

	/**
	 * Creates a clone of `dataView`.
	 *
	 * @private
	 * @param {Object} dataView The data view to clone.
	 * @param {boolean} [isDeep] Specify a deep clone.
	 * @returns {Object} Returns the cloned data view.
	 */
	function cloneDataView$1(dataView, isDeep) {
	  var buffer = isDeep ? cloneArrayBuffer$2(dataView.buffer) : dataView.buffer;
	  return new dataView.constructor(buffer, dataView.byteOffset, dataView.byteLength);
	}

	var _cloneDataView = cloneDataView$1;

	/** Used to match `RegExp` flags from their coerced string values. */

	var reFlags = /\w*$/;

	/**
	 * Creates a clone of `regexp`.
	 *
	 * @private
	 * @param {Object} regexp The regexp to clone.
	 * @returns {Object} Returns the cloned regexp.
	 */
	function cloneRegExp$1(regexp) {
	  var result = new regexp.constructor(regexp.source, reFlags.exec(regexp));
	  result.lastIndex = regexp.lastIndex;
	  return result;
	}

	var _cloneRegExp = cloneRegExp$1;

	var Symbol$1 = _Symbol;

	/** Used to convert symbols to primitives and strings. */
	var symbolProto = Symbol$1 ? Symbol$1.prototype : undefined,
	    symbolValueOf = symbolProto ? symbolProto.valueOf : undefined;

	/**
	 * Creates a clone of the `symbol` object.
	 *
	 * @private
	 * @param {Object} symbol The symbol object to clone.
	 * @returns {Object} Returns the cloned symbol object.
	 */
	function cloneSymbol$1(symbol) {
	  return symbolValueOf ? Object(symbolValueOf.call(symbol)) : {};
	}

	var _cloneSymbol = cloneSymbol$1;

	var cloneArrayBuffer$1 = _cloneArrayBuffer;

	/**
	 * Creates a clone of `typedArray`.
	 *
	 * @private
	 * @param {Object} typedArray The typed array to clone.
	 * @param {boolean} [isDeep] Specify a deep clone.
	 * @returns {Object} Returns the cloned typed array.
	 */
	function cloneTypedArray$1(typedArray, isDeep) {
	  var buffer = isDeep ? cloneArrayBuffer$1(typedArray.buffer) : typedArray.buffer;
	  return new typedArray.constructor(buffer, typedArray.byteOffset, typedArray.length);
	}

	var _cloneTypedArray = cloneTypedArray$1;

	var cloneArrayBuffer = _cloneArrayBuffer,
	    cloneDataView = _cloneDataView,
	    cloneRegExp = _cloneRegExp,
	    cloneSymbol = _cloneSymbol,
	    cloneTypedArray = _cloneTypedArray;

	/** `Object#toString` result references. */
	var boolTag$1 = '[object Boolean]',
	    dateTag$1 = '[object Date]',
	    mapTag$2 = '[object Map]',
	    numberTag$1 = '[object Number]',
	    regexpTag$1 = '[object RegExp]',
	    setTag$2 = '[object Set]',
	    stringTag$1 = '[object String]',
	    symbolTag$2 = '[object Symbol]';

	var arrayBufferTag$1 = '[object ArrayBuffer]',
	    dataViewTag$1 = '[object DataView]',
	    float32Tag$1 = '[object Float32Array]',
	    float64Tag$1 = '[object Float64Array]',
	    int8Tag$1 = '[object Int8Array]',
	    int16Tag$1 = '[object Int16Array]',
	    int32Tag$1 = '[object Int32Array]',
	    uint8Tag$1 = '[object Uint8Array]',
	    uint8ClampedTag$1 = '[object Uint8ClampedArray]',
	    uint16Tag$1 = '[object Uint16Array]',
	    uint32Tag$1 = '[object Uint32Array]';

	/**
	 * Initializes an object clone based on its `toStringTag`.
	 *
	 * **Note:** This function only supports cloning values with tags of
	 * `Boolean`, `Date`, `Error`, `Map`, `Number`, `RegExp`, `Set`, or `String`.
	 *
	 * @private
	 * @param {Object} object The object to clone.
	 * @param {string} tag The `toStringTag` of the object to clone.
	 * @param {boolean} [isDeep] Specify a deep clone.
	 * @returns {Object} Returns the initialized clone.
	 */
	function initCloneByTag$1(object, tag, isDeep) {
	  var Ctor = object.constructor;
	  switch (tag) {
	    case arrayBufferTag$1:
	      return cloneArrayBuffer(object);

	    case boolTag$1:
	    case dateTag$1:
	      return new Ctor(+object);

	    case dataViewTag$1:
	      return cloneDataView(object, isDeep);

	    case float32Tag$1: case float64Tag$1:
	    case int8Tag$1: case int16Tag$1: case int32Tag$1:
	    case uint8Tag$1: case uint8ClampedTag$1: case uint16Tag$1: case uint32Tag$1:
	      return cloneTypedArray(object, isDeep);

	    case mapTag$2:
	      return new Ctor;

	    case numberTag$1:
	    case stringTag$1:
	      return new Ctor(object);

	    case regexpTag$1:
	      return cloneRegExp(object);

	    case setTag$2:
	      return new Ctor;

	    case symbolTag$2:
	      return cloneSymbol(object);
	  }
	}

	var _initCloneByTag = initCloneByTag$1;

	var isObject$2 = isObject_1;

	/** Built-in value references. */
	var objectCreate = Object.create;

	/**
	 * The base implementation of `_.create` without support for assigning
	 * properties to the created object.
	 *
	 * @private
	 * @param {Object} proto The object to inherit from.
	 * @returns {Object} Returns the new object.
	 */
	var baseCreate$1 = (function() {
	  function object() {}
	  return function(proto) {
	    if (!isObject$2(proto)) {
	      return {};
	    }
	    if (objectCreate) {
	      return objectCreate(proto);
	    }
	    object.prototype = proto;
	    var result = new object;
	    object.prototype = undefined;
	    return result;
	  };
	}());

	var _baseCreate = baseCreate$1;

	var baseCreate = _baseCreate,
	    getPrototype = _getPrototype,
	    isPrototype = _isPrototype;

	/**
	 * Initializes an object clone.
	 *
	 * @private
	 * @param {Object} object The object to clone.
	 * @returns {Object} Returns the initialized clone.
	 */
	function initCloneObject$1(object) {
	  return (typeof object.constructor == 'function' && !isPrototype(object))
	    ? baseCreate(getPrototype(object))
	    : {};
	}

	var _initCloneObject = initCloneObject$1;

	var getTag$2 = _getTag,
	    isObjectLike$2 = isObjectLike_1;

	/** `Object#toString` result references. */
	var mapTag$1 = '[object Map]';

	/**
	 * The base implementation of `_.isMap` without Node.js optimizations.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a map, else `false`.
	 */
	function baseIsMap$1(value) {
	  return isObjectLike$2(value) && getTag$2(value) == mapTag$1;
	}

	var _baseIsMap = baseIsMap$1;

	var baseIsMap = _baseIsMap,
	    baseUnary$1 = _baseUnary,
	    nodeUtil$1 = _nodeUtil.exports;

	/* Node.js helper references. */
	var nodeIsMap = nodeUtil$1 && nodeUtil$1.isMap;

	/**
	 * Checks if `value` is classified as a `Map` object.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.3.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a map, else `false`.
	 * @example
	 *
	 * _.isMap(new Map);
	 * // => true
	 *
	 * _.isMap(new WeakMap);
	 * // => false
	 */
	var isMap$1 = nodeIsMap ? baseUnary$1(nodeIsMap) : baseIsMap;

	var isMap_1 = isMap$1;

	var getTag$1 = _getTag,
	    isObjectLike$1 = isObjectLike_1;

	/** `Object#toString` result references. */
	var setTag$1 = '[object Set]';

	/**
	 * The base implementation of `_.isSet` without Node.js optimizations.
	 *
	 * @private
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a set, else `false`.
	 */
	function baseIsSet$1(value) {
	  return isObjectLike$1(value) && getTag$1(value) == setTag$1;
	}

	var _baseIsSet = baseIsSet$1;

	var baseIsSet = _baseIsSet,
	    baseUnary = _baseUnary,
	    nodeUtil = _nodeUtil.exports;

	/* Node.js helper references. */
	var nodeIsSet = nodeUtil && nodeUtil.isSet;

	/**
	 * Checks if `value` is classified as a `Set` object.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.3.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a set, else `false`.
	 * @example
	 *
	 * _.isSet(new Set);
	 * // => true
	 *
	 * _.isSet(new WeakSet);
	 * // => false
	 */
	var isSet$1 = nodeIsSet ? baseUnary(nodeIsSet) : baseIsSet;

	var isSet_1 = isSet$1;

	var Stack = _Stack,
	    arrayEach = _arrayEach,
	    assignValue = _assignValue,
	    baseAssign = _baseAssign,
	    baseAssignIn = _baseAssignIn,
	    cloneBuffer = _cloneBuffer.exports,
	    copyArray = _copyArray,
	    copySymbols = _copySymbols,
	    copySymbolsIn = _copySymbolsIn,
	    getAllKeys = _getAllKeys,
	    getAllKeysIn = _getAllKeysIn,
	    getTag = _getTag,
	    initCloneArray = _initCloneArray,
	    initCloneByTag = _initCloneByTag,
	    initCloneObject = _initCloneObject,
	    isArray = isArray_1,
	    isBuffer = isBuffer$2.exports,
	    isMap = isMap_1,
	    isObject$1 = isObject_1,
	    isSet = isSet_1,
	    keys = keys_1,
	    keysIn = keysIn_1;

	/** Used to compose bitmasks for cloning. */
	var CLONE_DEEP_FLAG$1 = 1,
	    CLONE_FLAT_FLAG = 2,
	    CLONE_SYMBOLS_FLAG$1 = 4;

	/** `Object#toString` result references. */
	var argsTag = '[object Arguments]',
	    arrayTag = '[object Array]',
	    boolTag = '[object Boolean]',
	    dateTag = '[object Date]',
	    errorTag = '[object Error]',
	    funcTag = '[object Function]',
	    genTag = '[object GeneratorFunction]',
	    mapTag = '[object Map]',
	    numberTag = '[object Number]',
	    objectTag = '[object Object]',
	    regexpTag = '[object RegExp]',
	    setTag = '[object Set]',
	    stringTag = '[object String]',
	    symbolTag$1 = '[object Symbol]',
	    weakMapTag = '[object WeakMap]';

	var arrayBufferTag = '[object ArrayBuffer]',
	    dataViewTag = '[object DataView]',
	    float32Tag = '[object Float32Array]',
	    float64Tag = '[object Float64Array]',
	    int8Tag = '[object Int8Array]',
	    int16Tag = '[object Int16Array]',
	    int32Tag = '[object Int32Array]',
	    uint8Tag = '[object Uint8Array]',
	    uint8ClampedTag = '[object Uint8ClampedArray]',
	    uint16Tag = '[object Uint16Array]',
	    uint32Tag = '[object Uint32Array]';

	/** Used to identify `toStringTag` values supported by `_.clone`. */
	var cloneableTags = {};
	cloneableTags[argsTag] = cloneableTags[arrayTag] =
	cloneableTags[arrayBufferTag] = cloneableTags[dataViewTag] =
	cloneableTags[boolTag] = cloneableTags[dateTag] =
	cloneableTags[float32Tag] = cloneableTags[float64Tag] =
	cloneableTags[int8Tag] = cloneableTags[int16Tag] =
	cloneableTags[int32Tag] = cloneableTags[mapTag] =
	cloneableTags[numberTag] = cloneableTags[objectTag] =
	cloneableTags[regexpTag] = cloneableTags[setTag] =
	cloneableTags[stringTag] = cloneableTags[symbolTag$1] =
	cloneableTags[uint8Tag] = cloneableTags[uint8ClampedTag] =
	cloneableTags[uint16Tag] = cloneableTags[uint32Tag] = true;
	cloneableTags[errorTag] = cloneableTags[funcTag] =
	cloneableTags[weakMapTag] = false;

	/**
	 * The base implementation of `_.clone` and `_.cloneDeep` which tracks
	 * traversed objects.
	 *
	 * @private
	 * @param {*} value The value to clone.
	 * @param {boolean} bitmask The bitmask flags.
	 *  1 - Deep clone
	 *  2 - Flatten inherited properties
	 *  4 - Clone symbols
	 * @param {Function} [customizer] The function to customize cloning.
	 * @param {string} [key] The key of `value`.
	 * @param {Object} [object] The parent object of `value`.
	 * @param {Object} [stack] Tracks traversed objects and their clone counterparts.
	 * @returns {*} Returns the cloned value.
	 */
	function baseClone$1(value, bitmask, customizer, key, object, stack) {
	  var result,
	      isDeep = bitmask & CLONE_DEEP_FLAG$1,
	      isFlat = bitmask & CLONE_FLAT_FLAG,
	      isFull = bitmask & CLONE_SYMBOLS_FLAG$1;

	  if (customizer) {
	    result = object ? customizer(value, key, object, stack) : customizer(value);
	  }
	  if (result !== undefined) {
	    return result;
	  }
	  if (!isObject$1(value)) {
	    return value;
	  }
	  var isArr = isArray(value);
	  if (isArr) {
	    result = initCloneArray(value);
	    if (!isDeep) {
	      return copyArray(value, result);
	    }
	  } else {
	    var tag = getTag(value),
	        isFunc = tag == funcTag || tag == genTag;

	    if (isBuffer(value)) {
	      return cloneBuffer(value, isDeep);
	    }
	    if (tag == objectTag || tag == argsTag || (isFunc && !object)) {
	      result = (isFlat || isFunc) ? {} : initCloneObject(value);
	      if (!isDeep) {
	        return isFlat
	          ? copySymbolsIn(value, baseAssignIn(result, value))
	          : copySymbols(value, baseAssign(result, value));
	      }
	    } else {
	      if (!cloneableTags[tag]) {
	        return object ? value : {};
	      }
	      result = initCloneByTag(value, tag, isDeep);
	    }
	  }
	  // Check for circular references and return its corresponding clone.
	  stack || (stack = new Stack);
	  var stacked = stack.get(value);
	  if (stacked) {
	    return stacked;
	  }
	  stack.set(value, result);

	  if (isSet(value)) {
	    value.forEach(function(subValue) {
	      result.add(baseClone$1(subValue, bitmask, customizer, subValue, value, stack));
	    });
	  } else if (isMap(value)) {
	    value.forEach(function(subValue, key) {
	      result.set(key, baseClone$1(subValue, bitmask, customizer, key, value, stack));
	    });
	  }

	  var keysFunc = isFull
	    ? (isFlat ? getAllKeysIn : getAllKeys)
	    : (isFlat ? keysIn : keys);

	  var props = isArr ? undefined : keysFunc(value);
	  arrayEach(props || value, function(subValue, key) {
	    if (props) {
	      key = subValue;
	      subValue = value[key];
	    }
	    // Recursively populate clone (susceptible to call stack limits).
	    assignValue(result, key, baseClone$1(subValue, bitmask, customizer, key, value, stack));
	  });
	  return result;
	}

	var _baseClone = baseClone$1;

	var baseClone = _baseClone;

	/** Used to compose bitmasks for cloning. */
	var CLONE_DEEP_FLAG = 1,
	    CLONE_SYMBOLS_FLAG = 4;

	/**
	 * This method is like `_.clone` except that it recursively clones `value`.
	 *
	 * @static
	 * @memberOf _
	 * @since 1.0.0
	 * @category Lang
	 * @param {*} value The value to recursively clone.
	 * @returns {*} Returns the deep cloned value.
	 * @see _.clone
	 * @example
	 *
	 * var objects = [{ 'a': 1 }, { 'b': 2 }];
	 *
	 * var deep = _.cloneDeep(objects);
	 * console.log(deep[0] === objects[0]);
	 * // => false
	 */
	function cloneDeep(value) {
	  return baseClone(value, CLONE_DEEP_FLAG | CLONE_SYMBOLS_FLAG);
	}

	var cloneDeep_1 = cloneDeep;

	/*! *****************************************************************************
	Copyright (c) Microsoft Corporation.

	Permission to use, copy, modify, and/or distribute this software for any
	purpose with or without fee is hereby granted.

	THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
	REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
	AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
	INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
	LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
	OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
	PERFORMANCE OF THIS SOFTWARE.
	***************************************************************************** */

	function __awaiter(thisArg, _arguments, P, generator) {
	    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
	    return new (P || (P = Promise))(function (resolve, reject) {
	        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
	        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
	        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
	        step((generator = generator.apply(thisArg, _arguments || [])).next());
	    });
	}

	function __generator(thisArg, body) {
	    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
	    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
	    function verb(n) { return function (v) { return step([n, v]); }; }
	    function step(op) {
	        if (f) throw new TypeError("Generator is already executing.");
	        while (_) try {
	            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
	            if (y = 0, t) op = [op[0] & 2, t.value];
	            switch (op[0]) {
	                case 0: case 1: t = op; break;
	                case 4: _.label++; return { value: op[1], done: false };
	                case 5: _.label++; y = op[1]; op = [0]; continue;
	                case 7: op = _.ops.pop(); _.trys.pop(); continue;
	                default:
	                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
	                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
	                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
	                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
	                    if (t[2]) _.ops.pop();
	                    _.trys.pop(); continue;
	            }
	            op = body.call(thisArg, _);
	        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
	        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
	    }
	}

	/**
	 * Iterate through plugins and break in cases where the response is canceled. The
	 * response is canceled if component is updated or unmounted
	 * @param element {HTMLImageElement|HTMLVideoElement} Html Image or Video element
	 * @param pluginCloudinaryAsset {CloudinaryImage|CloudinaryVideo} The Cloudinary asset generated by base
	 * @param plugins {plugins} array of plugins passed in by the user
	 * @param pluginState {htmlPluginState} Holds cleanup callbacks and event subscriptions
	 */
	function render(element, pluginCloudinaryAsset, plugins, pluginState) {
	    return __awaiter(this, void 0, void 0, function () {
	        var i, response;
	        return __generator(this, function (_a) {
	            switch (_a.label) {
	                case 0:
	                    if (plugins === undefined)
	                        return [2 /*return*/];
	                    i = 0;
	                    _a.label = 1;
	                case 1:
	                    if (!(i < plugins.length)) return [3 /*break*/, 4];
	                    return [4 /*yield*/, plugins[i](element, pluginCloudinaryAsset, pluginState)];
	                case 2:
	                    response = _a.sent();
	                    if (response === 'canceled') {
	                        return [3 /*break*/, 4];
	                    }
	                    _a.label = 3;
	                case 3:
	                    i++;
	                    return [3 /*break*/, 1];
	                case 4: return [2 /*return*/];
	            }
	        });
	    });
	}

	var HtmlImageLayer = /** @class */ (function () {
	    function HtmlImageLayer(element, userCloudinaryImage, plugins) {
	        var _this = this;
	        this.imgElement = element;
	        this.htmlPluginState = { cleanupCallbacks: [], pluginEventSubscription: [] };
	        var pluginCloudinaryImage = cloneDeep_1(userCloudinaryImage);
	        render(element, pluginCloudinaryImage, plugins, this.htmlPluginState)
	            .then(function () {
	            _this.htmlPluginState.pluginEventSubscription.forEach(function (fn) { fn(); });
	            _this.imgElement.setAttribute('src', pluginCloudinaryImage.toURL());
	        });
	    }
	    /**
	     * Called when component is updated and re-triggers render
	     * @param userCloudinaryImage
	     * @param plugins
	     */
	    HtmlImageLayer.prototype.update = function (userCloudinaryImage, plugins) {
	        var _this = this;
	        var pluginCloudinaryImage = cloneDeep_1(userCloudinaryImage);
	        render(this.imgElement, pluginCloudinaryImage, plugins, this.htmlPluginState)
	            .then(function () {
	            _this.imgElement.setAttribute('src', pluginCloudinaryImage.toURL());
	        });
	    };
	    return HtmlImageLayer;
	}());

	/**
	 * @summary SDK
	 * @memberOf SDK
	 */
	class QualifierValue {
	    /**
	     *
	     * @param {QualifierValue | QualifierValue[] | any[] | string | number}qualifierValue
	     */
	    constructor(qualifierValue) {
	        this.values = [];
	        this.delimiter = ':'; // {value}{delimiter}{value}...
	        if (this.hasValue(qualifierValue)) {
	            this.addValue(qualifierValue);
	        }
	    }
	    /**
	     * @description Joins the provided values with the provided delimiter
	     */
	    toString() {
	        return this.values.join(this.delimiter);
	    }
	    /**
	     * @description Checks if the provided argument has a value
	     * @param {any} v
	     * @private
	     * @return {boolean}
	     */
	    hasValue(v) {
	        return typeof v !== 'undefined' && v !== null && v !== '';
	    }
	    /**
	     * @desc Adds a value for the this qualifier instance
	     * @param {any} value
	     * @return {this}
	     */
	    addValue(value) {
	        // Append value or array of values
	        if (Array.isArray(value)) {
	            this.values = this.values.concat(value);
	        }
	        else {
	            this.values.push(value);
	        }
	        // Remove falsy values
	        this.values = this.values.filter((v) => this.hasValue(v));
	        return this;
	    }
	    /**
	     * @description Sets the delimiter for this instance
	     * @param delimiter
	     */
	    setDelimiter(delimiter) {
	        this.delimiter = delimiter;
	        return this;
	    }
	}

	/**
	 * @summary SDK
	 * @memberOf SDK
	 */
	class Qualifier {
	    constructor(key, qualifierValue) {
	        this.delimiter = '_'; // {key}{delimiter}{qualifierValue}
	        this.key = key;
	        if (qualifierValue instanceof QualifierValue) {
	            this.qualifierValue = qualifierValue;
	        }
	        else {
	            this.qualifierValue = new QualifierValue();
	            this.qualifierValue.addValue(qualifierValue);
	        }
	    }
	    toString() {
	        const { key, delimiter, qualifierValue } = this;
	        return `${key}${delimiter}${qualifierValue.toString()}`;
	    }
	    addValue(value) {
	        this.qualifierValue.addValue(value);
	        return this;
	    }
	}

	/**
	 * @memberOf Qualifiers.Flag
	 * @extends {SDK.Qualifier}
	 * @description the FlagQualifier class
	 */
	class FlagQualifier extends Qualifier {
	    constructor(flagType, flagValue) {
	        let qualifierValue;
	        if (flagValue) {
	            qualifierValue = new QualifierValue([flagType, `${flagValue}`]).setDelimiter(':');
	        }
	        else {
	            qualifierValue = flagType;
	        }
	        super('fl', qualifierValue);
	    }
	    toString() {
	        return super.toString().replace(/\./, '%2E');
	    }
	}

	/**
	 * Sort a map by key
	 * @private
	 * @param map <string, any>
	 * @Return array of map's values sorted by key
	 */
	function mapToSortedArray(map, flags) {
	    const array = Array.from(map.entries());
	    // objects from the Array.from() method above are stored in array of arrays:
	    // [[qualifierKey, QualifierObj], [qualifierKey, QualifierObj]]
	    // Flags is an array of FlagQualifierObj
	    // We need to convert it to the same form: [flagQualifier] =>  ['fl', flagQualifier]
	    flags.forEach((flag) => {
	        array.push(['fl', flag]); // push ['fl', flagQualifier]
	    });
	    return array.sort().map((v) => v[1]);
	}

	/**
	 * @summary SDK
	 * @memberOf SDK
	 * @description Defines the category of transformation to perform.
	 */
	class Action {
	    constructor() {
	        // We're using map, to overwrite existing keys. for example:
	        // addParam(w_100).addQualifier(w_200) should result in w_200. and not w_100,w_200
	        this.qualifiers = new Map();
	        // Unlike regular qualifiers, there can be multiple flags in each url component /fl_1,fl_2/
	        // If the falgs are added to the qualifiers map, only a single flag could exist in a component (it's a map)
	        // So flags are stored separately until the very end because of that reason
	        this.flags = [];
	        this.delimiter = ','; // {qualifier}{delimiter}{qualifier} for example: `${'w_100'}${','}${'c_fill'}`
	        this.actionTag = ''; // A custom name tag to identify this action in the future
	    }
	    prepareQualifiers() { }
	    /**
	     * @description Returns the custom name tag that was given to this action
	     * @return {string}
	     */
	    getActionTag() {
	        return this.actionTag;
	    }
	    /**
	     * @description Sets the custom name tag for this action
	     * @return {this}
	     */
	    setActionTag(tag) {
	        this.actionTag = tag;
	        return this;
	    }
	    /**
	     * @description Calls toString() on all child qualifiers (implicitly by using .join()).
	     * @return {string}
	     */
	    toString() {
	        this.prepareQualifiers();
	        return mapToSortedArray(this.qualifiers, this.flags).join(this.delimiter);
	    }
	    /**
	     * @description Adds the parameter to the action.
	     * @param {SDK.Qualifier} qualifier
	     * @return {this}
	     */
	    addQualifier(qualifier) {
	        // if string, find the key and value
	        if (typeof qualifier === 'string') {
	            const [key, value] = qualifier.toLowerCase().split('_');
	            if (key === 'fl') {
	                // if string qualifier is a flag, store it in the flags arrays
	                this.flags.push(new FlagQualifier(value));
	            }
	            else {
	                // if the string qualifier is not a flag, create a new qualifier from it
	                this.qualifiers.set(key, new Qualifier(key, value));
	            }
	        }
	        else {
	            // if a qualifier object, insert to the qualifiers map
	            this.qualifiers.set(qualifier.key, qualifier);
	        }
	        return this;
	    }
	    /**
	     * @description Adds a flag to the current action.
	     * @param {Qualifiers.Flag} flag
	     * @return {this}
	     */
	    addFlag(flag) {
	        if (typeof flag === 'string') {
	            this.flags.push(new FlagQualifier(flag));
	        }
	        else {
	            if (flag instanceof FlagQualifier) {
	                this.flags.push(flag);
	            }
	        }
	        return this;
	    }
	    addValueToQualifier(qualifierKey, qualifierValue) {
	        this.qualifiers.get(qualifierKey).addValue(qualifierValue);
	        return this;
	    }
	}

	/**
	 * @description The Action class of the blur Builder.
	 * @extends SDK.Action
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class BlurAction extends Action {
	    constructor(strength) {
	        super();
	        this._strength = strength;
	    }
	    /**
	     * @description Specifies the region to blur.
	     * @param {NamedRegion} blurRegion
	     */
	    region(blurRegion) {
	        this._region = blurRegion;
	        return this;
	    }
	    /**
	     * @description Sets the strength of the blur effect.
	     * @param {number | string} strength
	     */
	    strength(strength) {
	        this._strength = strength;
	        return this;
	    }
	    prepareQualifiers() {
	        /*
	         * Blur with region is a unique object in this codebase.
	         * On top of Blur being an Action with Qualifiers,
	         * it also accepts a Qualifier called Region.
	         *
	         * This Qualifier is in itself composite of qualifiers (such as height, or width).
	         * The existence of Region changes the output of Blur in non traditional ways
	         * which forced this relatively ad-hoc implementation.
	         *
	         * Aside from all of that, all of the Qualifiers in the component should be alphabetized
	         * This happens naturally in the Action class,
	         * however since we're dealing with two levels of qualifiers (Blur and Region),
	         * these need to be merged.
	         *
	         * This function will merge the Region qualifiers with Blur
	         * and add all needed implicit qualifiers (like g_ocr_text).
	         * We're not using the full Gravity Qualifier here to prevent the code import for such a simplistic case
	         */
	        const str = this._strength ? `:${this._strength}` : '';
	        if ('_region' in this) {
	            const qualifiers = this._region.qualifiers;
	            // Copy qualifiers from the region "action" to the blur action
	            qualifiers.forEach((q) => this.addQualifier(q));
	            if (this._region.regionType === 'named') {
	                this.addQualifier(new Qualifier('e', `blur_region${str}`));
	            }
	            if (this._region.regionType === 'ocr_text') {
	                this.addQualifier(new Qualifier('e', `blur_region${str}`));
	                this.addQualifier(new Qualifier('g', `ocr_text`));
	            }
	            if (this._region.regionType === 'faces') {
	                this.addQualifier(new Qualifier('e', `blur_faces${str}`));
	            }
	        }
	        else {
	            this.addQualifier(new Qualifier('e', `blur${str}`));
	        }
	    }
	}

	/**
	 * @description A class that defines a simple effect of the type e_{effectName}
	 * @extends SDK.Action
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class SimpleEffectAction extends Action {
	    constructor(effectType, level) {
	        super();
	        const qualifierEffect = this.createEffectQualifier(effectType, level);
	        this.addQualifier(qualifierEffect);
	    }
	    createEffectQualifier(effectType, level) {
	        let qualifierValue;
	        if (level) {
	            qualifierValue = new QualifierValue([effectType, `${level}`]).setDelimiter(':');
	        }
	        else {
	            qualifierValue = new QualifierValue(effectType);
	        }
	        return new Qualifier('e', qualifierValue);
	    }
	}

	/**
	 * @description A base class for effects with a level, the extending class needs to implement a method that calls setLevel()
	 * @extends {Actions.Effect.SimpleEffectAction}
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class LeveledEffectAction extends SimpleEffectAction {
	    constructor(effectType, level) {
	        super(effectType, level);
	        this.effectType = effectType;
	    }
	    /**
	     *
	     * @description Sets the effect level for the action
	     * @param {string | number} level - The strength of the effect
	     * @protected
	     */
	    setLevel(level) {
	        const qualifierEffect = this.createEffectQualifier(this.effectType, level);
	        this.addQualifier(qualifierEffect);
	        return this;
	    }
	}

	/**
	 * Returns RGB or Color
	 * @private
	 * @param color
	 */
	function prepareColor(color) {
	    if (color) {
	        return color.match(/^#/) ? `rgb:${color.substr(1)}` : color;
	    }
	    else {
	        return color;
	    }
	}

	/**
	 * @description Vectorizes the image.
	 * @extends SDK.Action
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class VectorizeEffectAction extends Action {
	    constructor() {
	        super();
	    }
	    /**
	     * @description The number of colors. (Range: 2 to 30, Server default: 10)
	     * @param {number | string} num
	     * @return {this}
	     */
	    numOfColors(num) {
	        this._numOfColors = num;
	        return this;
	    }
	    /**
	     * @description The level of detail. Specify either a percentage of the original image (Range: 0.0 to 1.0) or an absolute number of pixels (Range: 0 to 1000). (Server default: 300)
	     * @param {number | string} num
	     * @return {this}
	     */
	    detailsLevel(num) {
	        this._detailsLevel = num;
	        return this;
	    }
	    /**
	     * @description The size of speckles to suppress. Specify either a percentage of the original image (Range: 0.0 to 1.0) or an absolute number of pixels (Range: 0 to 100, Server default: 2)
	     * @param {number | string} num
	     * @return {this}
	     */
	    despeckleLevel(num) {
	        this._despeckleLevel = num;
	        return this;
	    }
	    /**
	     * @description The corner threshold. Specify 100 for no smoothing (polygon corners), 0 for completely smooth corners. (Range: 0 to 100, Default: 25)
	     * @param {number | string} num
	     * @return {this}
	     */
	    cornersLevel(num) {
	        this._cornersLevel = num;
	        return this;
	    }
	    /**
	     * @description The optimization value. Specify 100 for least optimization and the largest file. (Range: 0 to 100, Server default: 100).
	     * @param {number} num
	     * @return {this}
	     */
	    paths(num) {
	        this._paths = num;
	        return this;
	    }
	    prepareQualifiers() {
	        let str = 'vectorize';
	        if (this._numOfColors) {
	            str += `:${new QualifierValue(`colors:${this._numOfColors}`).toString()}`;
	        }
	        if (this._detailsLevel) {
	            str += `:${new QualifierValue(`detail:${this._detailsLevel}`).toString()}`;
	        }
	        if (this._despeckleLevel) {
	            str += `:${new QualifierValue(`despeckle:${this._despeckleLevel}`).toString()}`;
	        }
	        if (this._paths) {
	            str += `:${new QualifierValue(`paths:${this._paths}`).toString()}`;
	        }
	        if (this._cornersLevel) {
	            str += `:${new QualifierValue(`corners:${this._cornersLevel}`).toString()}`;
	        }
	        this.addQualifier(new Qualifier('e', str));
	    }
	}

	/**
	 * @description A class that provides a built in level() method that sets the level of the effect
	 * @extends {Actions.Effect.LeveledEffectAction}
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class EffectActionWithLevel extends LeveledEffectAction {
	    level(value) {
	        return this.setLevel(value);
	    }
	}

	/**
	 * @description Applies stripes to the image to help people with common color-blind conditions to differentiate between colors that are similar for them.
	 *              You can replace colors using the xray() method.
	 * @extends SDK.Action
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class AssistColorBlindEffectAction extends Action {
	    constructor() {
	        super();
	        this.addQualifier(new Qualifier('e', new QualifierValue('assist_colorblind')));
	    }
	    /**
	     * @description Replaces problematic colors with colors that are easier to differentiate.
	     * @return {this}
	     */
	    xray() {
	        return this.addQualifier(new Qualifier('e', new QualifierValue(['assist_colorblind', 'xray']).setDelimiter(':')));
	    }
	    /**
	     * @description Applies stripes of the specified intensity to help people with common color blind conditions to differentiate between colors that are similar for them.
	     * @param {number | string} strength The intensity of the stripes. (Range: 1 to 100, Server default: 10)
	     * @return {this}
	     */
	    stripesStrength(strength) {
	        return this.addQualifier(new Qualifier('e', new QualifierValue(['assist_colorblind', strength]).setDelimiter(':')));
	    }
	}

	/**
	 * @description Applies a colorizing filter to the asset, use the methods in the class to adjust the filter
	 * @extends EffectActionWithLevel
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class ColorizeEffectAction extends EffectActionWithLevel {
	    /**
	     * @description The color to use for colorization. Specify HTML name or RGB hex code. (Server default: gray)
	     * @param {string} color HTML name(red, green, etc.) or RGB hex code. (Server default: gray)
	     * @return {this}
	     */
	    color(color) {
	        return this.addQualifier(new Qualifier('co', new QualifierValue(prepareColor(color))));
	    }
	}

	/**
	 * @description The Action class of the pixelate Builder
	 * @extends SDK.Action
	 * @memberOf Actions.Effect
	 * @see Visit {@link Actions.Effect|Effect} for an example
	 */
	class Pixelate extends Action {
	    constructor(squareSize) {
	        super();
	        this._squareSize = squareSize;
	    }
	    /**
	     * @description Specifies the region to piexlate.
	     * @param {NamedRegion} pixelateRegion
	     */
	    region(pixelateRegion) {
	        this._region = pixelateRegion;
	        return this;
	    }
	    /**
	     * @description Sets the squareSize of the pixelate effect.
	     * @param {number | string} squareSize
	     */
	    squareSize(squareSize) {
	        this._squareSize = squareSize;
	        return this;
	    }
	    prepareQualifiers() {
	        /*
	         * pixelate with region is a unique object in this codebase.
	         * On top of pixelate being an Action with Qualifiers,
	         * it also accepts a Qualifier called Region.
	         *
	         * This Qualifier is in itself composite of qualifiers (such as height, or width).
	         * The existence of Region changes the output of pixelate in non traditional ways
	         * which forced this relatively ad-hoc implementation.
	         *
	         * Aside from all of that, all of the Qualifiers in the component should be alphabetized
	         * This happens naturally in the Action class,
	         * however since we're dealing with two levels of qualifiers (pixelate and Region),
	         * these need to be merged.
	         *
	         * This function will merge the Region qualifiers with pixelate
	         * and add all needed implicit qualifiers (like g_ocr_text).
	         * We're not using the full Gravity Qualifier here to prevent the code import for such a simplistic case
	         */
	        const str = this._squareSize ? `:${this._squareSize}` : '';
	        if ('_region' in this) {
	            const qualifiers = this._region.qualifiers;
	            // Copy qualifiers from the region "action" to the pixelate action
	            qualifiers.forEach((q) => this.addQualifier(q));
	            if (this._region.regionType === 'named') {
	                this.addQualifier(new Qualifier('e', `pixelate_region${str}`));
	            }
	            if (this._region.regionType === 'ocr_text') {
	                this.addQualifier(new Qualifier('e', `pixelate_region${str}`));
	                this.addQualifier(new Qualifier('g', `ocr_text`));
	            }
	            if (this._region.regionType === 'faces') {
	                this.addQualifier(new Qualifier('e', `pixelate_faces${str}`));
	            }
	        }
	        else {
	            this.addQualifier(new Qualifier('e', `pixelate${str}`));
	        }
	    }
	}

	/**
	 * @summary action
	 * @description Applies a blurring filter to the asset.
	 * @memberOf Actions.Effect
	 * @param {number} blurLevel The strength of the blur. (Range: 1 to 2000, Server default: 100)
	 * @return {Actions.Effect.BlurAction}
	 */
	function blur(blurLevel) {
	    return new BlurAction(blurLevel);
	}
	/**
	 * @summary action
	 * @description Converts the image to gray-scale (multiple shades of gray).
	 * @memberOf Actions.Effect
	 * @return {Actions.Effect.SimpleEffectAction}
	 */
	function grayscale() {
	    return new SimpleEffectAction('grayscale');
	}
	/**
	 * @summary action
	 * @description Applies a colorizing filter to the asset.
	 * @memberOf Actions.Effect
	 * @param {number} colorizeLevel The strength of the color. (Range: 0 to 100, Server default: 100)
	 * @return {Actions.Effect.ColorizeEffectAction}
	 */
	function colorize(colorizeLevel) {
	    return new ColorizeEffectAction('colorize', colorizeLevel);
	}
	/**
	 * @summary action
	 * @description
	 * Vectorizes the image.
	 * Notes:
	 * To deliver the image as a vector image, make sure to change the format (or URL extension) to a vector format, such as SVG.</br>
	 * However, you can also deliver in a raster format if you just want to get the 'vectorized' graphic effect.</br>
	 * Large images are scaled down to 1000 pixels in the largest dimension before vectorization.
	 *
	 * @memberOf Actions.Effect
	 * @return {Actions.Effect.VectorizeEffectAction}
	 */
	function vectorize() {
	    return new VectorizeEffectAction();
	}
	/**
	 * @summary action
	 * @description
	 * Applies stripes to the image to help people with common color-blind conditions to differentiate between colors that are similar for them.</br>
	 * You can replace colors using the xRay() method of the \Cloudinary\Transformation\AssistColorBlind class.
	 * @memberOf Actions.Effect
	 * @return {Actions.Effect.AssistColorBlindEffectAction}
	 */
	function assistColorBlind() {
	    return new AssistColorBlindEffectAction();
	}
	/**
	 * @summary action
	 * @description Applies a pixelatering filter to the asset.
	 * @memberOf Actions.Effect
	 * @param {number} squareSize The squareSize in the pixelation. (Range: 1 to 2000, Server default: 100)
	 * @return {Actions.Effect.Pixelate}
	 */
	function pixelate(squareSize) {
	    return new Pixelate(squareSize);
	}

	/**
	 * @extends SDK.Action
	 * @description A class for background transformations.
	 */
	class BackgroundColor extends Action {
	    constructor(color) {
	        super();
	        this.addQualifier(new Qualifier('b', new QualifierValue(color).setDelimiter('_')));
	    }
	}

	/**
	 * @summary SDK
	 * @memberOf SDK
	 * @description Defines an action that's a string literal, no validations or manipulations are performed
	 */
	class RawAction {
	    constructor(raw) {
	        this.raw = raw;
	    }
	    toString() {
	        return this.raw;
	    }
	}

	/**
	 * @summary SDK
	 * @description - Defines how to transform an asset
	 * @memberOf SDK
	 */
	class Transformation {
	    constructor() {
	        this.actions = [];
	    }
	    /**
	     * @param {SDK.Action | string} action
	     * @return {this}
	     */
	    addAction(action) {
	        let actionToAdd;
	        if (typeof action === 'string') {
	            if (action.indexOf('/') >= 0) {
	                throw 'addAction cannot accept a string with a forward slash in it - /, use .addTransformation() instead';
	            }
	            else {
	                actionToAdd = new RawAction(action);
	            }
	        }
	        else {
	            actionToAdd = action;
	        }
	        this.actions.push(actionToAdd);
	        return this;
	    }
	    /**
	     * @description Allows the injection of a raw transformation as a string into the transformation, or a Transformation instance that was previously created
	     * @param {string | SDK.Transformation} tx
	     * @example
	     * import {Transformation} from "@cloudinary/url-gen";
	     *
	     * const transformation = new Transformation();
	     * transformation.addTransformation('w_100/w_200/w_300');
	     * @return {this}
	     */
	    addTransformation(tx) {
	        if (tx instanceof Transformation) {
	            // Concat the new actions into the existing actions
	            this.actions = this.actions.concat(tx.actions);
	        }
	        else {
	            this.actions.push(new RawAction(tx));
	        }
	        return this;
	    }
	    /**
	     * @return {string}
	     */
	    toString() {
	        return this.actions
	            .map((action) => {
	            return action.toString();
	        })
	            .filter((a) => a)
	            .join('/');
	    }
	    /**
	     * @description Delivers an animated GIF.
	     * @param {AnimatedAction} animatedAction
	     * @return {this}
	     */
	    animated(animatedAction) {
	        return this.addAction(animatedAction);
	    }
	    /**
	     * @description Adds a border around the image.
	     * @param {Border} borderAction
	     * @return {this}
	     */
	    border(borderAction) {
	        return this.addAction(borderAction);
	    }
	    /**
	     * @description Adjusts the shape of the delivered image. </br>
	     * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_transformations#image_shape_changes_and_distortion_effects|Shape changes and distortion effects}
	     * @param {IReshape} reshapeAction
	     * @return {this}
	     */
	    reshape(reshapeAction) {
	        return this.addAction(reshapeAction);
	    }
	    /**
	     * @description Resize the asset using provided resize action
	     * @param {ResizeSimpleAction} resizeAction
	     * @return {this}
	     */
	    resize(resizeAction) {
	        return this.addAction(resizeAction);
	    }
	    /**
	     * @param {DeliveryAction} quality
	     * @return {this}
	     */
	    quality(quality) {
	        return this.addAction(quality);
	    }
	    /**
	     * @description Rounds the specified corners of an image.
	     * @param roundCornersAction
	     * @return {this}
	     */
	    roundCorners(roundCornersAction) {
	        return this.addAction(roundCornersAction);
	    }
	    /**
	     * @description Adds an overlay over the base image.
	     * @param {LayerAction} overlayAction
	     * @return {this}
	     */
	    overlay(overlayAction) {
	        return this.addAction(overlayAction);
	    }
	    /**
	     * @description Adds an underlay under the base image.
	     * @param {LayerAction} underlayAction
	     * @return {this}
	     */
	    underlay(underlayAction) {
	        underlayAction.setLayerType('u');
	        return this.addAction(underlayAction);
	    }
	    /**
	     * @description Defines an new user variable.
	     * @param {VariableAction} variableAction
	     * @return {this}
	     */
	    addVariable(variableAction) {
	        return this.addAction(variableAction);
	    }
	    /**
	     * @description Specifies a condition to be met before applying a transformation.
	     * @param {ConditionalAction} conditionAction
	     * @return {this}
	     */
	    conditional(conditionAction) {
	        return this.addAction(conditionAction);
	    }
	    /**
	     * @description Applies a filter or an effect on an asset.
	     * @param {SimpleEffectAction} effectAction
	     * @return {this}
	     */
	    effect(effectAction) {
	        return this.addAction(effectAction);
	    }
	    /**
	     * @description Applies adjustment effect on an asset.
	     * @param action
	     * @return {this}
	     */
	    adjust(action) {
	        return this.addAction(action);
	    }
	    /**
	     * @description Rotates the asset by the given angle.
	     * @param {RotateAction} rotateAction
	     * @return {this}
	     */
	    rotate(rotateAction) {
	        return this.addAction(rotateAction);
	    }
	    /**
	     * @description Applies a pre-defined named transformation of the given name.
	     * @param {NamedTransformation} namedTransformation
	     * @return {this}
	     */
	    namedTransformation(namedTransformation) {
	        return this.addAction(namedTransformation);
	    }
	    /**
	     * @description Applies delivery action.
	     * @param deliveryAction
	     * @return {this}
	     */
	    delivery(deliveryAction) {
	        return this.addAction(deliveryAction);
	    }
	    /**
	     * @description Sets the color of the background.
	     * @param {Qualifiers.Color} color
	     * @return {this}
	     */
	    backgroundColor(color) {
	        return this.addAction(new BackgroundColor(prepareColor(color)));
	    }
	    /**
	     * @description Adds a layer in a Photoshop document.
	     * @param action
	     * @return {this}
	     */
	    psdTools(action) {
	        return this.addAction(action);
	    }
	    /**
	     * @description Extracts an image or a page using an index, a range, or a name from a layered media asset.
	     * @param action
	     * @return {this}
	     */
	    extract(action) {
	        return this.addAction(action);
	    }
	    /**
	     * @description Adds a flag as a separate action.
	     * @param {Qualifiers.Flag | string} flagQualifier
	     * @return {this}
	     */
	    addFlag(flagQualifier) {
	        const action = new Action();
	        let flagToAdd = flagQualifier;
	        if (typeof flagQualifier === 'string') {
	            flagToAdd = new FlagQualifier(flagQualifier);
	        }
	        action.addQualifier(flagToAdd);
	        return this.addAction(action);
	    }
	    /**
	     * @description Inject a custom function into the image transformation pipeline.
	     * @return {this}
	     */
	    customFunction(customFunction) {
	        return this.addAction(customFunction);
	    }
	    /**
	     * Transcodes the video (or audio) to another format.
	     * @param {Action} action
	     * @return {this}
	     */
	    transcode(action) {
	        return this.addAction(action);
	    }
	    /**
	     * Applies the specified video edit action.
	     *
	     * @param {videoEditType} action
	     * @return {this}
	     */
	    videoEdit(action) {
	        return this.addAction(action);
	    }
	}

	/**
	 * @description
	 * Returns a string representing the float value of the input, if the input was a number-like.
	 * Examples:
	 * - '1.0' -> '1.0'
	 * - 1 -> '1.0'
	 * - '5' -> '5.0'
	 * - 'auto' -> 'auto'
	 * @private
	 * @param {string|number} value
	 * @return {string}
	 */
	function toFloatAsString(value) {
	    // Turn the input to string
	    // The Function will return `returnValue` value if the input is not a number-like expression
	    const returnValue = value.toString();
	    // if the string contains letters, return the input
	    if (returnValue.match(/[A-Z]/gi)) {
	        return returnValue;
	    }
	    // If the leading digit is 0, and we have more than 1 digit, it's not a number.
	    // 00, 00000, 0x15 etc.
	    if (returnValue.length > 1 && returnValue[0] === '0') {
	        return returnValue;
	    }
	    // Final sanity check, parse the number as a float and check if its NaN
	    const isNumberLike = !isNaN(parseFloat(returnValue)) && returnValue.indexOf(':') === -1;
	    // If it's a number-like, but the input does not contain a decimal - add it.
	    if (isNumberLike && returnValue.indexOf('.') === -1) {
	        return `${returnValue}.0`;
	    }
	    else {
	        // If the input already contains a decimal, just return the value
	        return returnValue;
	    }
	}

	/**
	 * @memberOf Qualifiers.AspectRatio
	 * @extends {SDK.QualifierValue}
	 */
	class AspectRatioQualifierValue extends QualifierValue {
	}

	/**
	 * @description Defines flags that you can use to alter the default transformation behavior.
	 * @namespace Flag
	 * @memberOf Qualifiers
	 */
	/**
	 * @summary qualifier
	 * @memberOf Qualifiers.Flag
	 * @description Automatically use lossy compression when delivering animated GIF files.
	 *
	 * This flag can also be used as a conditional flag for delivering PNG files: it tells Cloudinary to deliver the
	 * image in PNG format (as requested) unless there is no transparency channel - in which case deliver in JPEG
	 * format.
	 * @return {Qualifiers.Flag.FlagQualifier}
	 */
	function lossy() {
	    return new FlagQualifier('lossy');
	}
	/**
	 * @summary qualifier
	 * @memberOf Qualifiers.Flag
	 * @description When used with automatic fetch_format (f_auto): ensures that images with a transparency channel will be
	 * delivered in PNG format.
	 * @return {Qualifiers.Flag.FlagQualifier}
	 */
	function preserveTransparency() {
	    return new FlagQualifier('preserve_transparency');
	}
	/**
	 * @summary qualifier
	 * @memberOf Qualifiers.Flag
	 * @description Generates a JPG image using the progressive (interlaced) JPG format.
	 *
	 * This format allows the browser to quickly show a low-quality rendering of the image until the full-quality
	 * image is loaded.
	 *
	 * @param {string} mode The mode to determine a specific progressive outcome as follows:
	 * * semi - A smart optimization of the decoding time, compression level and progressive rendering
	 *          (less iterations). This is the default mode when using q_auto.
	 * * steep - Delivers a preview very quickly, and in a single later phase improves the image to
	 *           the required resolution.
	 * * none  - Use this to deliver a non-progressive image. This is the default mode when setting
	 *           a specific value for quality.
	 * @return {Qualifiers.Flag.FlagQualifier}
	 */
	function progressive(mode) {
	    return new FlagQualifier('progressive', mode);
	}
	/**
	 * @summary qualifier
	 * @memberOf Qualifiers.Flag
	 * @description Modifies percentage-based width & height parameters of overlays and underlays (e.g., 1.0) to be relative to the overlaid region
	 * @return {Qualifiers.Flag.FlagQualifier}
	 */
	function regionRelative() {
	    return new FlagQualifier('region_relative');
	}
	/**
	 * @summary qualifier
	 * @memberOf Qualifiers.Flag
	 * @description Modifies percentage-based width & height parameters of overlays and underlays (e.g., 1.0) to be relative to the containing image instead of the added layer.
	 * @return {Qualifiers.Flag.FlagQualifier}
	 */
	function relative() {
	    return new FlagQualifier('relative');
	}

	/**
	 * @description Defines a resize using width and height.
	 * @extends SDK.Action
	 * @memberOf Actions.Resize
	 * @see Visit {@link Actions.Resize| Resize} for examples
	 */
	class ResizeSimpleAction extends Action {
	    /**
	     * @param {string} cropType
	     * @param {number | string} cropWidth The required width of a transformed asset.
	     * @param {number | string} cropHeight The required height of a transformed asset.
	     */
	    constructor(cropType, cropWidth, cropHeight) {
	        super();
	        if (cropWidth) {
	            this.addQualifier(new Qualifier('w', cropWidth));
	        }
	        if (cropHeight) {
	            this.addQualifier(new Qualifier('h', cropHeight));
	        }
	        this.addQualifier(new Qualifier('c', cropType));
	    }
	    /**
	     * @description Sets the height of the resize
	     * @param {string | number} x The height in pixels (if an integer is specified) or as a percentage (if a float is specified).
	     */
	    height(x) {
	        return this.addQualifier(new Qualifier('h', x));
	    }
	    /**
	     * @description Sets the width of the resize
	     * @param {string | number} x The width in pixels (if an integer is specified) or as a percentage (if a float is specified).
	     */
	    width(x) {
	        return this.addQualifier(new Qualifier('w', x));
	    }
	    /**
	     * @description Sets the aspect ratio of the asset.
	     * For a list of supported types see {@link Qualifiers.AspectRatio |
	      * AspectRatio values}
	     * @param {AspectRatioType|number|string} ratio The new aspect ratio, specified as a percentage or ratio.
	     * @return {this}
	     */
	    aspectRatio(ratio) {
	        // toFloatAsString is used to ensure 1 turns into 1.0
	        if (ratio instanceof AspectRatioQualifierValue) {
	            return this.addQualifier(new Qualifier('ar', ratio));
	        }
	        if (typeof ratio === 'number' || typeof ratio === 'string') {
	            return this.addQualifier(new Qualifier('ar', toFloatAsString(ratio)));
	        }
	        if (ratio instanceof FlagQualifier) {
	            return this.addFlag(ratio);
	        }
	    }
	    /**
	     * @description Modifies percentage-based width & height parameters of overlays and underlays (e.g., 1.0) to be relative to the containing image instead of the added layer.
	     * @return {this}
	     */
	    relative() {
	        return this.addFlag(relative());
	    }
	    /**
	     * @description Modifies percentage-based width & height parameters of overlays and underlays (e.g., 1.0) to be relative to the overlaid region
	     * @return {this}
	     */
	    regionRelative() {
	        return this.addFlag(regionRelative());
	    }
	}
	var ResizeSimpleAction$1 = ResizeSimpleAction;

	/**
	 * @description Defines an advanced resize.
	 * @extends Actions.Resize.ResizeSimpleAction
	 * @memberOf Actions.Resize
	 * @see Visit {@link Actions.Resize| Resize} for examples
	 */
	class ResizeAdvancedAction extends ResizeSimpleAction$1 {
	    /**
	     * @description Which part of the original image to include.
	     * @param {Qualifiers.Gravity} gravity
	     */
	    gravity(gravity) {
	        if (typeof gravity === "string") {
	            return this.addQualifier(new Qualifier('g', gravity));
	        }
	        return this.addQualifier(gravity);
	    }
	}
	var ResizeAdvancedAction$1 = ResizeAdvancedAction;

	/**
	 * @description Defines an advanced resize with padding.
	 * @extends Actions.Resize.ResizeAdvancedAction
	 * @memberOf Actions.Resize
	 * @see Visit {@link Actions.Resize| Resize} for examples
	 */
	class ResizePadAction extends ResizeAdvancedAction$1 {
	    /**
	     * @description Sets the background.
	     * @param {Qualifiers.Background} backgroundQualifier Defines the background color to use instead of
	     * transparent background areas or when resizing with padding.
	     */
	    background(backgroundQualifier) {
	        return this.addQualifier(backgroundQualifier);
	    }
	    gravity(direction) {
	        return this.addQualifier(direction);
	    }
	    /**
	     * @description Horizontal position for custom-coordinates based padding.
	     * @param {number} x The x position.
	     */
	    offsetX(x) {
	        return this.addQualifier(new Qualifier('x', x));
	    }
	    /**
	     * @description Vertical position for custom-coordinates based padding
	     * @param {number} y The y position.
	     */
	    offsetY(y) {
	        return this.addQualifier(new Qualifier('y', y));
	    }
	}
	var ResizePadAction$1 = ResizePadAction;

	/**
	 * @memberOf Gravity.GravityQualifier
	 * @extends {SDK.Qualifier}
	 */
	class GravityQualifier extends Qualifier {
	    /**
	     * @param value, an array containing (GravityObject | AutoGravity | string) or a string;
	     */
	    constructor(value) {
	        super('g', new QualifierValue(value));
	    }
	}

	/**
	 * @description Defines a scaling resize action.
	 * @extends Actions.Resize.ResizeSimpleAction
	 * @memberOf Actions.Resize
	 * @see Visit {@link Actions.Resize| Resize} for examples
	 */
	class ScaleAction extends ResizeSimpleAction$1 {
	    /**
	     * @description Changes the aspect ratio of an image while retaining all important content and avoiding unnatural
	     * distortions.
	     * @return {this}
	     */
	    liquidRescaling() {
	        return this.addQualifier(new GravityQualifier('liquid'));
	    }
	}

	/**
	 * @description Defines how to crop an asset
	 * @extends Actions.Resize.ResizeAdvancedAction
	 * @memberOf Actions.Resize
	 * @see Visit {@link Actions.Resize| Resize} for examples
	 */
	class ResizeCropAction extends ResizeAdvancedAction$1 {
	    /**
	     * @description Horizontal position for custom-coordinates based cropping.
	     * @param {number} x The x position.
	     */
	    x(x) {
	        return this.addQualifier(new Qualifier('x', x));
	    }
	    /**
	     * @description Vertical position for custom-coordinates based cropping
	     * @param {number} y The y position.
	     */
	    y(y) {
	        return this.addQualifier(new Qualifier('y', y));
	    }
	    /**
	     * @description Controls how much of the original image surrounding the face to keep when using either the 'crop' or 'thumb' cropping modes with face detection.
	     * @param {number | string} z The zoom factor. (Default: 1.0)
	     */
	    zoom(z) {
	        return this.addQualifier(new Qualifier('z', z));
	    }
	}
	var ResizeCropAction$1 = ResizeCropAction;

	/**
	 * @description Defines how to crop-fill an asset
	 * @extends Actions.Resize.ResizeAdvancedAction
	 * @memberOf Actions.Resize
	 * @see Visit {@link Actions.Resize| Resize} for examples
	 */
	class ResizeFillAction extends ResizeAdvancedAction$1 {
	    /**
	     * @description Absolute X position when used with Gravity.xyCenter {@link Qualifiers.Gravity.GravityQualifier}}
	     * @param {number} x The x position.
	     */
	    x(x) {
	        return this.addQualifier(new Qualifier('x', x));
	    }
	    /**
	     * @description Absolute Y position when used with Gravity.xyCenter {@link Qualifiers.Gravity.GravityQualifier}}
	     * @param {number} y The y position.
	     */
	    y(y) {
	        return this.addQualifier(new Qualifier('y', y));
	    }
	}

	/**
	 * @description Determines how to crop, scale, and/or zoom the delivered asset according to the requested dimensions.
	 * @memberOf Actions
	 * @namespace Resize
	 * @see Learn more about Gravity and Focus {@link Qualifiers.Gravity| here }
	 * @example
	 * <caption> <h4>Scaling an image</h4> </caption>
	 * import {Cloudinary} from "@cloudinary/url-gen";
	 *
	 * const yourCldInstance = new Cloudinary({cloud:{cloudName:'demo'}});
	 * const image = yourCldInstance.image('woman');
	 *
	 * const {scale, fit, pad, crop} from '@cloudinary/url-gen/actions/resize';
	 *
	 * image.resize( scale(100, 100) );
	 * // All resize actions have a similar interface.
	 * // image.resize( fit(100, 100)) );
	 * // image.resize( pad(100, 100)) );
	 * // image.resize( crop(100, 100)) );
	 * // However, Some actions have additional arguments exposed as builder methods.
	 * // See the documentation for each method for more information
	 *
	 *
	 * // Alternative syntax, using builder methods
	 * image.resize(
	 *  scale()
	 *    .width(100)
	 *    .height(100)
	 * );
	 * image.toURL()
	 *
	 * @example
	 * <caption> <h4>Cropping with automatic focus(Gravity)</h4> </caption>
	 * import {Cloudinary} from "@cloudinary/url-gen";
	 *
	 * const yourCldInstance = new Cloudinary({cloud:{cloudName:'demo'}});
	 * const image = yourCldInstance.image('woman');
	 *
	 * import {scale} from '@cloudinary/url-gen/actions/resize';
	 * import {autoGravity} from '@cloudinary/url-gen/qualifiers/gravity';
	 *
	 * image.resize( crop(100, 100).gravity(autoGravity()) );
	 *
	 * // Alternative syntax, using builder methods
	 * image.resize(
	 *  scale()
	 *    .width(100)
	 *    .height(100)
	 *    .gravity(autoGravity())
	 * );
	 * image.toURL()
	 */
	/**
	 * @summary action
	 * @description
	 * Changes the size of the image exactly to the given width and height without necessarily retaining the original aspect ratio:<br/>
	 * all original image parts are visible but might be stretched or shrunk.
	 * @memberOf Actions.Resize
	 * @param {number|string} width The required width of a transformed asset.
	 * @param {number|string} height The required height of a transformed asset.
	 * @return {Actions.Resize.ScaleAction}
	 */
	function scale(width, height) {
	    return new ScaleAction('scale', width, height);
	}
	/**
	 * @summary action
	 * @description Extracts a region of the given width and height out of the original image.
	 * @memberOf Actions.Resize
	 * @param {number|string} width The required width of a transformed asset.
	 * @param {number|string} height The required height of a transformed asset.
	 * @return {Actions.Resize.ResizeCropAction}
	 */
	function crop(width, height) {
	    return new ResizeCropAction$1('crop', width, height);
	}
	/**
	 * @summary action
	 * @description
	 * Creates an image with the exact given width and height without distorting the image.<br/>
	 * This option first scales up or down as much as needed to at least fill both of the given dimensions.<br/><br/>
	 * If the requested aspect ratio is different than the original, cropping will occur on the dimension that exceeds the requested size after scaling.
	 * @memberOf Actions.Resize
	 * @param {number|string} width The required width of a transformed asset.
	 * @param {number|string} height The required height of a transformed asset.
	 * @return {Actions.Resize.ResizeFillAction}
	 */
	function fill(width, height) {
	    return new ResizeFillAction('fill', width, height);
	}
	/**
	 * @summary action
	 * @description
	 * Resizes the asset to fill the given width and height while retaining the original aspect ratio.
	 *
	 * If the proportions of the original asset do not match the given width and height, padding is added to the asset
	 * to reach the required size.
	 * @memberOf Actions.Resize
	 * @param {number|string} width The required width of a transformed asset.
	 * @param {number|string} height The required height of a transformed asset.
	 * @return {Actions.Resize.ResizePadAction}
	 */
	function pad(width, height) {
	    return new ResizePadAction$1('pad', width, height);
	}

	/**
	 * @description Defines the visual appearance of the background.
	 * @memberOf Qualifiers.Background
	 * @extends {SDK.Qualifier}
	 */
	class BackgroundQualifier extends Qualifier {
	    constructor(backgroundValue) {
	        // The qualifier key for this qualifier
	        super('b');
	        // Such as color (b_red)
	        if (backgroundValue) {
	            this.addValue(backgroundValue);
	        }
	    }
	}

	/**
	 * @description Defines the background color to use when resizing with padding.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BackgroundQualifier}
	 */
	class BaseCommonBackground extends BackgroundQualifier {
	    constructor() {
	        super();
	        this._palette = [];
	    }
	    /**
	     * @description Selects the strongest contrasting color to use for padding.
	     * @return {this}
	     */
	    contrast() {
	        this._constrast = true;
	        return this;
	    }
	    /**
	     * @description Defines the custom colors to use when resizing using content-aware padding.
	     * @param {...string} colors One or more colors - Example: palette('green', 'red', blue')
	     * @return {this}
	     */
	    palette(...colors) {
	        this._palette = colors.map((color) => {
	            return prepareColor(color);
	        });
	        return this;
	    }
	}

	/**
	 * @description Automatically determines the color to use for padding, if needed when resizing an asset. Selects the
	 * predominant color from the border of the image.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BaseCommonBackground}
	 */
	class BackgroundAutoBorderQualifier extends BaseCommonBackground {
	    /**
	     * @description
	     * Stringify the qualifier
	     * BackgroundQualifiers don't have a value, but instead override the toString() function.
	     */
	    toString() {
	        return `
    b_auto:border
    ${this._constrast ? '_contrast' : ''}
    ${this._palette.length ? `:palette_${this._palette.join('_')}` : ''}
    `.replace(/\s+/g, '');
	    }
	}

	/**
	 * @description Defines the gradient fade effect to use for the background when resizing with padding.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BaseCommonBackground}
	 */
	class BaseGradientBackground extends BaseCommonBackground {
	    /**
	     *
	     * @description Sets the number of predominant colors to use (2 or 4).
	     * @param {number} num
	     * @return {this}
	     */
	    gradientColors(num) {
	        this._gradientColors = num;
	        return this;
	    }
	    /**
	     * @description Sets the direction for a background gradient fade effect.
	     * @param {Qualifiers.GradientDirection | GradientDirectionType | string} direction Use one of these functions
	     * provided by {@link Qualifiers.GradientDirection|GradientDirection}
	     * @return {this}
	     */
	    gradientDirection(direction) {
	        this._gradientDirection = direction;
	        return this;
	    }
	}

	/**
	 * @description Specifies that the gradient fade effect, used for the background when resizing with padding, uses the
	 * predominant colors in the border pixels of the image.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BaseGradientBackground}
	 */
	class BackgroundBorderGradientQualifier extends BaseGradientBackground {
	    /**
	     * @description
	     * Stringify the qualifier
	     * BackgroundQualifiers don't have a value, but instead override the toString() function.
	     */
	    toString() {
	        return `
    b_auto:border_gradient
    ${this._constrast ? '_contrast' : ''}
    ${this._gradientColors ? `:${this._gradientColors}` : ''}
    ${this._gradientDirection ? `:${this._gradientDirection}` : ''}
    ${this._palette.length ? `:palette_${this._palette.join('_')}` : ''}
    `.replace(/\s+/g, '');
	    }
	}

	/**
	 * @description Automatically determines the color to use for padding, if needed when resizing an asset. Selects the
	 * predominant color from the whole image.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BaseCommonBackground}
	 */
	class BackgroundAutoPredominantQualifier extends BaseCommonBackground {
	    /**
	     * @description
	     * Stringify the qualifier
	     * BackgroundQualifiers don't have a value, but instead override the toString() function.
	     */
	    toString() {
	        return `
    b_auto:predominant
    ${this._constrast ? '_contrast' : ''}
    ${this._palette.length ? `:palette_${this._palette.join('_')}` : ''}
    `.replace(/\s+/g, '');
	    }
	}

	/**
	 * @description Specifies that the gradient fade effect, used for the background when resizing with padding, uses the
	 * predominant colors in the whole of the image.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BaseGradientBackground}
	 */
	class BackgroundPredominantGradientQualifier extends BaseGradientBackground {
	    /**
	     * @description
	     * Stringify the qualifier
	     * BackgroundQualifiers don't have a value, but instead override the toString() function.
	     */
	    toString() {
	        return `
    b_auto:predominant_gradient
    ${this._constrast ? '_contrast' : ''}
    ${this._gradientColors ? `:${this._gradientColors}` : ''}
    ${this._gradientDirection ? `:${this._gradientDirection}` : ''}
    ${this._palette.length ? `:palette_${this._palette.join('_')}` : ''}
    `.replace(/\s+/g, '');
	    }
	}

	/**
	 * @description A class for blurred background transformations.
	 * @memberOf Qualifiers.Background
	 * @extends {Qualifiers.Background.BackgroundQualifier}
	 */
	class BlurredBackgroundAction extends BackgroundQualifier {
	    /**
	     * @description Sets the intensity of the blur.
	     * @param {number} value - The intensity of the blur.
	     */
	    intensity(value) {
	        this.intensityLevel = value;
	        return this;
	    }
	    /**
	     * @description Sets the brightness of the background.
	     * @param {number} value - The brightness of the background.
	     */
	    brightness(value) {
	        this.brightnessLevel = value;
	        return this;
	    }
	    /**
	     * @description
	     * Stringify the qualifier
	     * BackgroundQualifiers don't have a value, but instead override the toString() function
	     */
	    toString() {
	        // b_blurred:{intensity}:{brightness}
	        return `
    b_blurred
    ${this.intensityLevel ? `:${this.intensityLevel}` : ''}
    ${this.brightnessLevel ? `:${this.brightnessLevel}` : ''}
    `.replace(/\s+/g, '');
	    }
	}
	var BlurredBackgroundAction$1 = BlurredBackgroundAction;

	/**
	 * @description Defines the background color to use instead of transparent background areas or when resizing with padding.
	 *
	 * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_transformations#setting_background_color|Setting background for images} | {@link https://cloudinary.com/documentation/video_manipulation_and_delivery#background_color|Setting background for videos}
	 *
	 * @namespace Background
	 * @memberOf Qualifiers
	 */
	/**
	 * @summary qualifier
	 * @description Selects the predominant color while taking only the image border pixels into account.
	 * @memberOf Qualifiers.Background
	 * @return {BackgroundAutoBorderQualifier}
	 */
	function border() {
	    return new BackgroundAutoBorderQualifier();
	}
	/**
	 * @summary qualifier
	 * @description Automatically determines the color to use for padding, if needed when resizing an asset.
	 *
	 * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_transformations#content_aware_padding|Content-aware padding}
	 * @memberOf Qualifiers.Background
	 * @return {Qualifiers.Background.BackgroundQualifier}
	 */
	function auto$1() {
	    return new BackgroundQualifier('auto');
	}
	/**
	 * @summary qualifier
	 * @description Applies a padding gradient fade effect using the predominant colors in the border of the image.
	 * @memberOf Qualifiers.Background
	 * @return {BackgroundBorderGradientQualifier}
	 */
	function borderGradient() {
	    return new BackgroundBorderGradientQualifier();
	}
	/**
	 * @summary qualifier
	 * @description Applies a padding gradient fade effect using the predominant colors in the image.
	 * @memberOf Qualifiers.Background
	 * @return {BackgroundPredominantGradientQualifier}
	 */
	function predominantGradient() {
	    return new BackgroundPredominantGradientQualifier();
	}
	/**
	 * @summary qualifier
	 * @description Selects the predominant color while taking all pixels in the image into account
	 * @memberOf Qualifiers.Background
	 * @return {BackgroundAutoPredominantQualifier}
	 */
	function predominant() {
	    return new BackgroundAutoPredominantQualifier();
	}
	/**
	 * @summary qualifier
	 * @description Selects the predominant color while taking all pixels in the image into account.
	 * @memberOf Qualifiers.Background
	 * @return {Qualifiers.Background.BackgroundQualifier}
	 */
	function color(colorStr) {
	    return new BackgroundQualifier(prepareColor(colorStr));
	}
	/**
	 * @summary qualifier
	 * @description Selects the predominant color while taking all pixels in the image into account.
	 * @memberOf Qualifiers.Background
	 * @return {BlurredBackgroundAction}
	 */
	function blurred() {
	    return new BlurredBackgroundAction$1();
	}
	const Background = {
	    auto: auto$1,
	    border: border,
	    borderGradient: borderGradient,
	    predominantGradient: predominantGradient,
	    predominant: predominant,
	    color: color,
	    blurred: blurred
	};

	/**
	 * @description The class for the CompassGravity builder
	 * @memberOf Qualifiers.Gravity
	 * @extends {Qualifiers.Gravity.GravityQualifier}
	 */
	class CompassGravity extends GravityQualifier {
	    constructor(dir) {
	        // Required due to https://github.com/microsoft/TypeScript/issues/13029
	        /* istanbul ignore next */
	        super(dir);
	    }
	}

	/**
	 * @description Defines the gravity based on directional values from a compass.
	 * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_transformations#control_gravity | Control gravity for images}
	 * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_transformations#control_gravity | Control gravity for videos}
	 * @param {Qualifiers.Compass} direction A compass Values
	 * @memberOf Qualifiers.Gravity
	 * @example
	 * import {Cloudinary} from "@cloudinary/url-gen";
	 * import {compass} from "@cloudinary/url-gen/qualifiers/gravity";
	 * import {north} from "@cloudinary/url-gen/qualifiers/compass";
	 * import {crop} from "@cloudinary/url-gen/actions/resize";
	 *
	 * const yourCldInstance = new Cloudinary({cloud: {cloudName: 'demo'}});
	 * const image = yourCldInstance.image('woman');
	 * image.resize(crop().width(300).gravity(compass(north())))
	 * @return {CompassGravity}
	 */
	function compass(direction) {
	    return new CompassGravity(direction);
	}

	/**
	 * @memberOf Qualifiers.Compass
	 * @extends {SDK.QualifierValue}
	 */
	class CompassQualifier extends QualifierValue {
	    constructor(val) {
	        super();
	        this.val = val;
	    }
	    toString() {
	        return this.val;
	    }
	}

	/**
	 * @summary qualifier
	 * @memberOf Qualifiers.Compass
	 * @description North east corner (top right).
	 * @return {Qualifiers.Compass.CompassQualifier} Compass
	 */
	function northEast() {
	    return new CompassQualifier('north_east');
	}

	/**
	 * @description Qualifies the delivery of an asset.
	 * @memberOf Actions.Delivery
	 * @extends SDK.Action
	 */
	class DeliveryAction extends Action {
	    /**
	     * @param {string} deliveryKey A generic Delivery Action Key (such as q, f, dn, etc.)
	     * @param {string} deliveryType A Format Qualifiers for the action, such as Quality.auto()
	     * @see Visit {@link Actions.Delivery|Delivery} for an example
	     */
	    constructor(deliveryKey, deliveryType) {
	        super();
	        this.addQualifier(new Qualifier(deliveryKey, deliveryType));
	    }
	}

	/**
	 * @description Contains functions to select the mode when using a progressive format.
	 * <b>Learn more</b>: {@link https://cloudinary.com/documentation/transformation_reference#fl_progressive | Progressive modes}
	 * @memberOf Qualifiers
	 * @namespace Progressive
	 * @example
	 * import {Cloudinary} from "@cloudinary/url-gen";
	 * import {format} from "@cloudinary/url-gen/actions/delivery";
	 * import {jpg} from "@cloudinary/url-gen/qualifiers/format";
	 * import {steep} from "@cloudinary/url-gen/qualifiers/progressive";
	 *
	 * const yourCldInstance = new Cloudinary({cloud: {cloudName: 'demo'}});
	 * const image = yourCldInstance.image('woman');
	 * image.delivery(format(jpg()).progressive(steep()))
	 */
	class ProgressiveQualifier extends FlagQualifier {
	    constructor(mode) {
	        super('progressive', mode);
	    }
	}

	/**
	 * @memberOf Actions.Delivery
	 * @extends {Actions.Delivery.DeliveryAction}
	 * @see Visit {@link Actions.Delivery|Delivery} for an example
	 */
	class DeliveryFormat extends DeliveryAction {
	    /**
	     * @description Uses lossy compression when delivering animated GIF files.
	     * @return {this}
	     */
	    lossy() {
	        this.addFlag(lossy());
	        return this;
	    }
	    /**
	     * @description Uses progressive compression when delivering JPG file format.
	     * @return {this}
	     */
	    progressive(mode) {
	        if (mode instanceof ProgressiveQualifier) {
	            this.addFlag(mode);
	        }
	        else {
	            this.addFlag(progressive(mode));
	        }
	        return this;
	    }
	    /**
	     * @description Ensures that images with a transparency channel are delivered in PNG format.
	     */
	    preserveTransparency() {
	        this.addFlag(preserveTransparency());
	        return this;
	    }
	}

	/**
	 * @description Controls the quality of the delivered image or video.
	 * @memberOf Actions.Delivery
	 * @extends {Actions.Delivery.DeliveryAction}
	 * @see Visit {@link Actions.Delivery|Delivery} for an example
	 */
	class DeliveryQualityAction extends DeliveryAction {
	    /**
	     * @param {Qualifiers.Quality} qualityValue a Quality value
	     */
	    constructor(qualityValue) {
	        super('q', qualityValue.toString());
	        this.qualityValue = qualityValue;
	    }
	    /**
	     * Selet the Chroma sub sampling</br>
	     * <b>Learn more</b>: {@link https://cloudinary.com/documentation/image_transformations#toggling_chroma_subsampling | Toggling chroma subsampling}
	     * @param {420 | 444 | number} type The chroma sub sampling type
	     */
	    chromaSubSampling(type) {
	        const qualityWithSubSampling = new QualifierValue([this.qualityValue, type]);
	        qualityWithSubSampling.setDelimiter(':');
	        // We either have chroma or quantization, but not both
	        return this.addQualifier(new Qualifier('q', qualityWithSubSampling));
	    }
	    /**
	     * Controls the final quality by setting a maximum quantization percentage
	     * @param {number} val
	     */
	    quantization(val) {
	        const qualityWithQuantization = new QualifierValue([this.qualityValue, `qmax_${val}`]);
	        qualityWithQuantization.setDelimiter(':');
	        // We either have chroma or quantization, but not both
	        return this.addQualifier(new Qualifier('q', qualityWithQuantization));
	    }
	}

	/**
	 * @description Defines transformations for delivering your assets without changing the visual or audio experience for the end user.
	 * @memberOf Actions
	 * @namespace Delivery
	 * @example
	 * See the examples under every method
	 */
	/**
	 * @summary action
	 * @description Defines the format of the delivered asset.
	 *
	 * <b>Learn more:</b>
	 * {@link https://cloudinary.com/documentation/image_transformations#image_format_support | Image formats}
	 * {@link https://cloudinary.com/documentation/video_manipulation_and_delivery#transcoding_video_to_other_formats | Video formats}
	 *
	 * @memberOf Actions.Delivery
	 * @param {string} format The file format. For a list of supported format types see {@link Qualifiers.Format| format types} for
	 * possible values
	 * @return {Actions.Delivery.DeliveryFormat}
	 * @example
	 * import {Cloudinary} from "@cloudinary/url-gen";
	 * import {format} from "@cloudinary/url-gen/actions/delivery";
	 *
	 * const yourCldInstance = new Cloudinary({cloud:{cloudName:'demo'}});
	 * const image = yourCldInstance.image('woman');
	 * image.delivery(
	 *  format('jpg'),
	 * );
	 *
	 */
	function format(format) {
	    return new DeliveryFormat('f', format);
	}
	/**
	 * @summary action
	 * @description Controls the quality of the delivered image or video.
	 *
	 * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_optimization#how_to_optimize_image_quality | Image quality}
	 *  {@link https://cloudinary.com/documentation/video_manipulation_and_delivery#quality_control | Video quality}
	 * @memberOf Actions.Delivery
	 * @param {QualityTypes | string | number | Qualifiers.Quality} qualityType For a list of supported quality types see
	 * {@link Qualifiers.Quality| quality types} for
	 * possible values.
	 * @return {Actions.Delivery.DeliveryQualityAction}
	 * @example
	 * import {Cloudinary} from "@cloudinary/url-gen";
	 * import {quality} from "@cloudinary/url-gen/actions/delivery";
	 * import {quality} from "@cloudinary/url-gen/qualifiers/quantity";
	 *
	 * const yourCldInstance = new Cloudinary({cloud:{cloudName:'demo'}});
	 * const image = yourCldInstance.image('woman');
	 * image.delivery(
	 *  quality('auto'),
	 * );
	 */
	function quality(qualityType) {
	    return new DeliveryQualityAction(qualityType);
	}

	/**
	 * @memberOf Qualifiers.Format
	 * @extends {SDK.QualifierValue}
	 */
	class FormatQualifier extends QualifierValue {
	    constructor(val) {
	        super(val);
	    }
	}

	/**
	 * @summary qualifier
	 * @description Image format svg.
	 * @memberOf Qualifiers.Format
	 * @return {Qualifiers.Format.FormatQualifier}
	 */
	function svg() { return new FormatQualifier('svg'); }
	/**
	 * @summary qualifier
	 * @description Image format auto.
	 * @memberOf Qualifiers.Format
	 * @return {Qualifiers.Format.FormatQualifier}
	 */
	function auto() { return new FormatQualifier('auto'); }

	/**
	 * Predefined accessibility transformations
	 * @const {Object} Cloudinary.ACCESSIBILITY_MODES
	 */
	var ACCESSIBILITY_MODES = {
	    'darkmode': colorize(70).color('black'),
	    'brightmode': colorize(40).color('white'),
	    'monochrome': grayscale(),
	    'colorblind': assistColorBlind()
	};
	/**
	 * Predefined vectorize placeholder transformation
	 */
	var VECTORIZE = new Transformation()
	    .effect(vectorize())
	    .delivery(quality('auto'))
	    .delivery(format(svg()));
	/**
	 * Predefined pixelate placeholder transformation
	 */
	var PIXELATE = new Transformation()
	    .effect(pixelate())
	    .delivery(quality('auto'))
	    .delivery(format(auto()));
	/**
	 * Predefined blur placeholder transformation
	 */
	var BLUR = new Transformation()
	    .effect(blur(2000))
	    .delivery(quality('auto'))
	    .delivery(format(auto()));
	/**
	 * Predefined predominant color placeholder transformation
	 */
	var PREDOMINANT_COLOR_TRANSFORM = new Transformation()
	    .resize(pad('iw_div_2').aspectRatio(1).background(Background.auto()))
	    .resize(crop(1, 1).gravity(compass(northEast())))
	    .resize(fill().height('ih').width('iw'))
	    .delivery(quality('auto'))
	    .delivery(format(auto()));
	/**
	 * Predefined placeholder image options
	 */
	var PLACEHOLDER_IMAGE_OPTIONS = {
	    'vectorize': VECTORIZE,
	    'pixelate': PIXELATE,
	    'blur': BLUR,
	    'predominant-color': PREDOMINANT_COLOR_TRANSFORM
	};
	/**
	 * transparent gif
	 */
	var singleTransparentPixel = 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';
	/**
	 * Convert common video file extensions to mime types
	 * Most other common video file extensions have an identical mime type so do not need conversion.
	 */
	var VIDEO_MIME_TYPES = {
	    'flv': 'x-flv',
	    '3gp': '3gpp',
	    'mov': 'quicktime',
	    'mpg': 'mpeg',
	    'avi': 'x-msvideo',
	    'wmv': 'x-ms-wmv',
	    'ogv': 'ogg'
	};

	var HtmlVideoLayer = /** @class */ (function () {
	    function HtmlVideoLayer(element, userCloudinaryVideo, sources, plugins, videoAttributes) {
	        var _this = this;
	        this.mimeType = 'video';
	        this.mimeSubTypes = VIDEO_MIME_TYPES;
	        this.videoElement = element;
	        this.originalVideo = userCloudinaryVideo;
	        this.htmlPluginState = { cleanupCallbacks: [], pluginEventSubscription: [] };
	        var pluginCloudinaryVideo = cloneDeep_1(userCloudinaryVideo);
	        render(element, userCloudinaryVideo, plugins, this.htmlPluginState)
	            .then(function () {
	            _this.htmlPluginState.pluginEventSubscription.forEach(function (fn) { fn(); });
	            _this.setVideoAttributes(videoAttributes);
	            _this.handleSourceToVideo(pluginCloudinaryVideo, sources);
	        });
	    }
	    /**
	     * Handles user supplied sources or default sources
	     * @param userCloudinaryVideo {CloudinaryVideo}
	     * @param sources
	     */
	    HtmlVideoLayer.prototype.handleSourceToVideo = function (userCloudinaryVideo, sources) {
	        var _this = this;
	        // checks if user supplied sources
	        if (sources) {
	            this.generateUserSources(userCloudinaryVideo, sources);
	        }
	        else {
	            var defaultTypes = ['webm', 'mp4', 'ogg'];
	            defaultTypes.forEach(function (type) {
	                _this.appendSourceTag(userCloudinaryVideo, type);
	            });
	        }
	    };
	    /**
	     * Generate sources based on user input
	     * @param userCloudinaryVideo {CloudinaryVideo}
	     * @param sources
	     */
	    HtmlVideoLayer.prototype.generateUserSources = function (userCloudinaryVideo, sources) {
	        var _this = this;
	        sources.map(function (_a) {
	            var type = _a.type, codecs = _a.codecs, transcode = _a.transcode;
	            return (_this.appendSourceTag(cloneDeep_1(userCloudinaryVideo)
	                .transcode(transcode), type, _this.buildMimeType(type, codecs)));
	        });
	    };
	    /**
	     * Appends source tag to html video element
	     * @param userCloudinaryVideo {CloudinaryVideo}
	     * @param type {string}
	     * @param mimeType {string}
	     */
	    HtmlVideoLayer.prototype.appendSourceTag = function (userCloudinaryVideo, type, mimeType) {
	        var source = document.createElement('source');
	        source.src = userCloudinaryVideo.toURL() + "." + type;
	        source.type = mimeType ? mimeType : "video/" + type;
	        this.videoElement.appendChild(source);
	    };
	    /**
	     * Determines MIME type of given source type and codecs.
	     * @param type - format of the video
	     * @param codecs - optional information about codecs of the video
	     */
	    HtmlVideoLayer.prototype.buildMimeType = function (type, codecs) {
	        var mimeType = this.mimeType + "/" + (this.mimeSubTypes[type] || type);
	        if (codecs) {
	            mimeType += "; codecs=" + (Array.isArray(codecs) ? codecs.join(', ') : codecs);
	        }
	        return mimeType;
	    };
	    /**
	     * Iterates through the video attributes and sets to true if passed in by the user.
	     * In case of poster, sets the poster.
	     * @param videoAttributes {object} Supported attributes: controls, loop, muted, poster, preload, autoplay, playsinline
	     */
	    HtmlVideoLayer.prototype.setVideoAttributes = function (videoAttributes) {
	        if (!videoAttributes)
	            return;
	        for (var _i = 0, _a = Object.entries(videoAttributes); _i < _a.length; _i++) {
	            var _b = _a[_i], key = _b[0], value = _b[1];
	            if (value) {
	                if (key === 'poster') {
	                    this.videoElement.poster = value;
	                }
	                else {
	                    this.videoElement[key] = true;
	                }
	            }
	        }
	    };
	    /**
	     * Called when component is updated. If our video source has changed, a video reload is triggered.
	     * @param updatedCloudinaryVideo
	     * @param sources
	     * @param plugins
	     * @param videoAttributes
	     */
	    HtmlVideoLayer.prototype.update = function (updatedCloudinaryVideo, sources, plugins, videoAttributes) {
	        var _this = this;
	        if (updatedCloudinaryVideo !== this.originalVideo) {
	            var sourcesToDelete = this.videoElement.getElementsByTagName("SOURCE");
	            while (sourcesToDelete[0])
	                sourcesToDelete[0].parentNode.removeChild(sourcesToDelete[0]);
	            render(this.videoElement, updatedCloudinaryVideo, plugins, this.htmlPluginState)
	                .then(function () {
	                _this.setVideoAttributes(videoAttributes);
	                _this.handleSourceToVideo(updatedCloudinaryVideo, sources);
	                _this.videoElement.load();
	            });
	        }
	    };
	    return HtmlVideoLayer;
	}());

	/**
	 * Predefined screen dimension table ordered by ranking
	 */
	var screeWidths = [
	    1366,
	    828,
	    1920,
	    3840,
	    1536,
	    750,
	    1280,
	    1600,
	    1440
	];

	var DEFAULT_MAX_WIDTH = 3840;
	var DEFAULT_MIN_WIDTH = 375;
	// constant used to generate the number of default breakpoints
	var MAX_IMAGES = 5;
	// constant used to estimate and simulate dpr
	var DEFAULT_DPR_THRESHOLD = 768;
	var HtmlPictureSourceLayer = /** @class */ (function () {
	    function HtmlPictureSourceLayer(image, minWidth, maxWidth, sizes, autoOptimalBreakpoints, breakpoints, relativeWidth) {
	        if (relativeWidth === void 0) { relativeWidth = 1; }
	        this.image = image;
	        this.minWidth = minWidth;
	        this.maxWidth = maxWidth;
	        this.sizes = sizes;
	        this.relativeWidth = relativeWidth;
	        this.autoOptimalBreakpoints = autoOptimalBreakpoints;
	        this.breakpoints = breakpoints;
	    }
	    HtmlPictureSourceLayer.prototype.toTag = function () {
	        var source = document.createElement('source');
	        this.setMedia(source);
	        this.setSize(source);
	        this.generateSrcsetUrl(source);
	        return source;
	    };
	    /**
	     * @description Sets Source element's media attribute
	     * @param source {HTMLSourceElement} Source element
	     */
	    HtmlPictureSourceLayer.prototype.setMedia = function (source) {
	        if (this.maxWidth && this.minWidth) {
	            source.media = "(min-width: " + this.minWidth + "px) and (max-width: " + this.maxWidth + "px)";
	        }
	        else if (this.maxWidth) {
	            source.media = "(max-width: " + this.maxWidth + "px)";
	        }
	        else if (this.minWidth) {
	            source.media = "(min-width: " + this.minWidth + "px)";
	        }
	    };
	    /**
	     * @description Sets Source element's size attribute
	     * @param source {HTMLSourceElement} Source element
	     */
	    HtmlPictureSourceLayer.prototype.setSize = function (source) {
	        if (this.sizes && this.autoOptimalBreakpoints) {
	            console.error("Invalid input: sizes must not be used with autoOptimalBreakpoints");
	        }
	        if (this.sizes) {
	            source.sizes = this.sizes;
	        }
	        else if (this.relativeWidth && this.autoOptimalBreakpoints) {
	            source.sizes = Math.ceil(this.relativeWidth * 100) + "vw";
	        }
	    };
	    /**
	     * @description Generates Source element srcset attribute
	     * @param source {HTMLSourceElement} Source element
	     */
	    HtmlPictureSourceLayer.prototype.generateSrcsetUrl = function (source) {
	        var _this = this;
	        var breakpoints = this.getBreakPoints();
	        if (breakpoints.length === 0) {
	            source.srcset = this.image.toURL();
	            return;
	        }
	        var breakpointUrls = [];
	        breakpoints.forEach(function (breakPoint) {
	            var width = Math.ceil(breakPoint * _this.relativeWidth);
	            var clone = cloneDeep_1(_this.image);
	            breakpointUrls.push(clone.resize(scale(width)).toURL() + " " + width + "w");
	        });
	        source.srcset = breakpointUrls.join(',');
	    };
	    /**
	      * @description An array of breakpoints is calculated and returned based on the
	      * predefined constants in screenWidths
	      * See conditions bellow
	      */
	    HtmlPictureSourceLayer.prototype.getBreakPoints = function () {
	        if (this.breakpoints) {
	            return this.breakpoints;
	        }
	        if (!this.autoOptimalBreakpoints) {
	            return [];
	        }
	        if (!this.minWidth && !this.maxWidth) {
	            console.error("Invalid input: either maxWidth or minWidth is required");
	        }
	        if (this.minWidth > this.maxWidth) {
	            console.error("Invalid input: maxWidth must be greater than minWidth");
	        }
	        var _a = this.getPhysicalDimension(), physicalMinWidth = _a[0], physicalMaxWidth = _a[1];
	        if (physicalMinWidth === physicalMaxWidth) {
	            return [physicalMaxWidth];
	        }
	        //Take 5 (or up to 5) sorted widths with highest rank between physicalMaxWidth and physicalMinWidth
	        return screeWidths.filter(function (width) {
	            return width <= physicalMaxWidth && width >= physicalMinWidth;
	        }).slice(0, MAX_IMAGES)
	            .sort(function (a, b) { return a - b; });
	    };
	    /**
	     * @description Returns the physical dimensions
	     */
	    HtmlPictureSourceLayer.prototype.getPhysicalDimension = function () {
	        var _a;
	        var maxWidth = this.maxWidth || DEFAULT_MAX_WIDTH;
	        var minWidth = this.minWidth || DEFAULT_MIN_WIDTH;
	        var physicalMaxWidth = this.getDprDimension(maxWidth);
	        var physicalMinWidth = this.getDprDimension(minWidth);
	        // if physicalMinWidth is greater we swap
	        if (physicalMinWidth > physicalMaxWidth) {
	            _a = [physicalMaxWidth, physicalMinWidth], physicalMinWidth = _a[0], physicalMaxWidth = _a[1];
	        }
	        return [physicalMinWidth, physicalMaxWidth];
	    };
	    /**
	     * @description If dimension is less than default DPR threshold we multiply by 2
	     * @param dimension {number} Estimates DPR bases on max and min width
	     */
	    HtmlPictureSourceLayer.prototype.getDprDimension = function (dimension) {
	        return dimension < DEFAULT_DPR_THRESHOLD ? dimension * 2 : dimension;
	    };
	    return HtmlPictureSourceLayer;
	}());

	var HtmlPictureLayer = /** @class */ (function () {
	    function HtmlPictureLayer(element, userCloudinaryImage, breakpoints, sources, autoOptimalBreakpoints, relativeWidth) {
	        var _this = this;
	        if (relativeWidth === void 0) { relativeWidth = 1; }
	        this.pictureElement = element;
	        sources.forEach(function (source) { _this.setSourceTag(breakpoints, source, autoOptimalBreakpoints, relativeWidth); });
	        this.setImageTag(userCloudinaryImage);
	    }
	    /**
	     * @description Sets image tag based on user's Cloudinary image input
	     * @param userCloudinaryImage {CloudinaryImage} Cloudinary image instance
	     */
	    HtmlPictureLayer.prototype.setImageTag = function (userCloudinaryImage) {
	        var imageTag = document.createElement('img');
	        imageTag.src = userCloudinaryImage.toURL();
	        this.pictureElement.appendChild(imageTag);
	    };
	    /**
	     * @description Calls HtmlPictureSourceLayer to set the source tag bases on user's input
	     * @param breakpoints {number[]} User's breakpoint array.
	     * @param sourceInput {PictureSource} Source input passed in by user.
	     * @param autoOptimalBreakpoints {boolean} boolean to turn on auto optimal breakpoint generation. Default set to
	     * false
	     * @param relativeWidth {number} Floating point number from: 0 to 1. Default set to 1
	     */
	    HtmlPictureLayer.prototype.setSourceTag = function (breakpoints, sourceInput, autoOptimalBreakpoints, relativeWidth) {
	        var sourceTest = new HtmlPictureSourceLayer(sourceInput.image, sourceInput.minWidth, sourceInput.maxWidth, sourceInput.sizes, autoOptimalBreakpoints, breakpoints, relativeWidth)
	            .toTag();
	        this.pictureElement.appendChild(sourceTest);
	    };
	    /**
	     * @description Called when component is updated.
	     * @param updatedCloudinaryImage {CloudinaryImage} Updated Cloudinary image instance
	     * @param breakpoints
	     * @param sources {HTMLSourceElement} Source element
	     * @param autoOptimalBreakpoints {boolean} boolean to turn on auto optimal breakpoint generation. Default set to
	     * @param relativeWidth {number} Floating point number from: 0 to 1. Default set to 1
	     */
	    HtmlPictureLayer.prototype.update = function (updatedCloudinaryImage, breakpoints, sources, autoOptimalBreakpoints, relativeWidth) {
	        var _this = this;
	        var sourcesToDelete = this.pictureElement.getElementsByTagName("SOURCE");
	        while (sourcesToDelete[0])
	            sourcesToDelete[0].parentNode.removeChild(sourcesToDelete[0]);
	        this.setImageTag(updatedCloudinaryImage);
	        sources.forEach(function (source) { _this.setSourceTag(breakpoints, source, autoOptimalBreakpoints, relativeWidth); });
	    };
	    return HtmlPictureLayer;
	}());

	/**
	 * lodash (Custom Build) <https://lodash.com/>
	 * Build: `lodash modularize exports="npm" -o ./`
	 * Copyright jQuery Foundation and other contributors <https://jquery.org/>
	 * Released under MIT license <https://lodash.com/license>
	 * Based on Underscore.js 1.8.3 <http://underscorejs.org/LICENSE>
	 * Copyright Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
	 */

	/** Used as the `TypeError` message for "Functions" methods. */
	var FUNC_ERROR_TEXT = 'Expected a function';

	/** Used as references for various `Number` constants. */
	var NAN = 0 / 0;

	/** `Object#toString` result references. */
	var symbolTag = '[object Symbol]';

	/** Used to match leading and trailing whitespace. */
	var reTrim = /^\s+|\s+$/g;

	/** Used to detect bad signed hexadecimal string values. */
	var reIsBadHex = /^[-+]0x[0-9a-f]+$/i;

	/** Used to detect binary string values. */
	var reIsBinary = /^0b[01]+$/i;

	/** Used to detect octal string values. */
	var reIsOctal = /^0o[0-7]+$/i;

	/** Built-in method references without a dependency on `root`. */
	var freeParseInt = parseInt;

	/** Detect free variable `global` from Node.js. */
	var freeGlobal = typeof commonjsGlobal == 'object' && commonjsGlobal && commonjsGlobal.Object === Object && commonjsGlobal;

	/** Detect free variable `self`. */
	var freeSelf = typeof self == 'object' && self && self.Object === Object && self;

	/** Used as a reference to the global object. */
	var root = freeGlobal || freeSelf || Function('return this')();

	/** Used for built-in method references. */
	var objectProto = Object.prototype;

	/**
	 * Used to resolve the
	 * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
	 * of values.
	 */
	var objectToString = objectProto.toString;

	/* Built-in method references for those with the same name as other `lodash` methods. */
	var nativeMax = Math.max,
	    nativeMin = Math.min;

	/**
	 * Gets the timestamp of the number of milliseconds that have elapsed since
	 * the Unix epoch (1 January 1970 00:00:00 UTC).
	 *
	 * @static
	 * @memberOf _
	 * @since 2.4.0
	 * @category Date
	 * @returns {number} Returns the timestamp.
	 * @example
	 *
	 * _.defer(function(stamp) {
	 *   console.log(_.now() - stamp);
	 * }, _.now());
	 * // => Logs the number of milliseconds it took for the deferred invocation.
	 */
	var now = function() {
	  return root.Date.now();
	};

	/**
	 * Creates a debounced function that delays invoking `func` until after `wait`
	 * milliseconds have elapsed since the last time the debounced function was
	 * invoked. The debounced function comes with a `cancel` method to cancel
	 * delayed `func` invocations and a `flush` method to immediately invoke them.
	 * Provide `options` to indicate whether `func` should be invoked on the
	 * leading and/or trailing edge of the `wait` timeout. The `func` is invoked
	 * with the last arguments provided to the debounced function. Subsequent
	 * calls to the debounced function return the result of the last `func`
	 * invocation.
	 *
	 * **Note:** If `leading` and `trailing` options are `true`, `func` is
	 * invoked on the trailing edge of the timeout only if the debounced function
	 * is invoked more than once during the `wait` timeout.
	 *
	 * If `wait` is `0` and `leading` is `false`, `func` invocation is deferred
	 * until to the next tick, similar to `setTimeout` with a timeout of `0`.
	 *
	 * See [David Corbacho's article](https://css-tricks.com/debouncing-throttling-explained-examples/)
	 * for details over the differences between `_.debounce` and `_.throttle`.
	 *
	 * @static
	 * @memberOf _
	 * @since 0.1.0
	 * @category Function
	 * @param {Function} func The function to debounce.
	 * @param {number} [wait=0] The number of milliseconds to delay.
	 * @param {Object} [options={}] The options object.
	 * @param {boolean} [options.leading=false]
	 *  Specify invoking on the leading edge of the timeout.
	 * @param {number} [options.maxWait]
	 *  The maximum time `func` is allowed to be delayed before it's invoked.
	 * @param {boolean} [options.trailing=true]
	 *  Specify invoking on the trailing edge of the timeout.
	 * @returns {Function} Returns the new debounced function.
	 * @example
	 *
	 * // Avoid costly calculations while the window size is in flux.
	 * jQuery(window).on('resize', _.debounce(calculateLayout, 150));
	 *
	 * // Invoke `sendMail` when clicked, debouncing subsequent calls.
	 * jQuery(element).on('click', _.debounce(sendMail, 300, {
	 *   'leading': true,
	 *   'trailing': false
	 * }));
	 *
	 * // Ensure `batchLog` is invoked once after 1 second of debounced calls.
	 * var debounced = _.debounce(batchLog, 250, { 'maxWait': 1000 });
	 * var source = new EventSource('/stream');
	 * jQuery(source).on('message', debounced);
	 *
	 * // Cancel the trailing debounced invocation.
	 * jQuery(window).on('popstate', debounced.cancel);
	 */
	function debounce(func, wait, options) {
	  var lastArgs,
	      lastThis,
	      maxWait,
	      result,
	      timerId,
	      lastCallTime,
	      lastInvokeTime = 0,
	      leading = false,
	      maxing = false,
	      trailing = true;

	  if (typeof func != 'function') {
	    throw new TypeError(FUNC_ERROR_TEXT);
	  }
	  wait = toNumber(wait) || 0;
	  if (isObject(options)) {
	    leading = !!options.leading;
	    maxing = 'maxWait' in options;
	    maxWait = maxing ? nativeMax(toNumber(options.maxWait) || 0, wait) : maxWait;
	    trailing = 'trailing' in options ? !!options.trailing : trailing;
	  }

	  function invokeFunc(time) {
	    var args = lastArgs,
	        thisArg = lastThis;

	    lastArgs = lastThis = undefined;
	    lastInvokeTime = time;
	    result = func.apply(thisArg, args);
	    return result;
	  }

	  function leadingEdge(time) {
	    // Reset any `maxWait` timer.
	    lastInvokeTime = time;
	    // Start the timer for the trailing edge.
	    timerId = setTimeout(timerExpired, wait);
	    // Invoke the leading edge.
	    return leading ? invokeFunc(time) : result;
	  }

	  function remainingWait(time) {
	    var timeSinceLastCall = time - lastCallTime,
	        timeSinceLastInvoke = time - lastInvokeTime,
	        result = wait - timeSinceLastCall;

	    return maxing ? nativeMin(result, maxWait - timeSinceLastInvoke) : result;
	  }

	  function shouldInvoke(time) {
	    var timeSinceLastCall = time - lastCallTime,
	        timeSinceLastInvoke = time - lastInvokeTime;

	    // Either this is the first call, activity has stopped and we're at the
	    // trailing edge, the system time has gone backwards and we're treating
	    // it as the trailing edge, or we've hit the `maxWait` limit.
	    return (lastCallTime === undefined || (timeSinceLastCall >= wait) ||
	      (timeSinceLastCall < 0) || (maxing && timeSinceLastInvoke >= maxWait));
	  }

	  function timerExpired() {
	    var time = now();
	    if (shouldInvoke(time)) {
	      return trailingEdge(time);
	    }
	    // Restart the timer.
	    timerId = setTimeout(timerExpired, remainingWait(time));
	  }

	  function trailingEdge(time) {
	    timerId = undefined;

	    // Only invoke if we have `lastArgs` which means `func` has been
	    // debounced at least once.
	    if (trailing && lastArgs) {
	      return invokeFunc(time);
	    }
	    lastArgs = lastThis = undefined;
	    return result;
	  }

	  function cancel() {
	    if (timerId !== undefined) {
	      clearTimeout(timerId);
	    }
	    lastInvokeTime = 0;
	    lastArgs = lastCallTime = lastThis = timerId = undefined;
	  }

	  function flush() {
	    return timerId === undefined ? result : trailingEdge(now());
	  }

	  function debounced() {
	    var time = now(),
	        isInvoking = shouldInvoke(time);

	    lastArgs = arguments;
	    lastThis = this;
	    lastCallTime = time;

	    if (isInvoking) {
	      if (timerId === undefined) {
	        return leadingEdge(lastCallTime);
	      }
	      if (maxing) {
	        // Handle invocations in a tight loop.
	        timerId = setTimeout(timerExpired, wait);
	        return invokeFunc(lastCallTime);
	      }
	    }
	    if (timerId === undefined) {
	      timerId = setTimeout(timerExpired, wait);
	    }
	    return result;
	  }
	  debounced.cancel = cancel;
	  debounced.flush = flush;
	  return debounced;
	}

	/**
	 * Checks if `value` is the
	 * [language type](http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types)
	 * of `Object`. (e.g. arrays, functions, objects, regexes, `new Number(0)`, and `new String('')`)
	 *
	 * @static
	 * @memberOf _
	 * @since 0.1.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is an object, else `false`.
	 * @example
	 *
	 * _.isObject({});
	 * // => true
	 *
	 * _.isObject([1, 2, 3]);
	 * // => true
	 *
	 * _.isObject(_.noop);
	 * // => true
	 *
	 * _.isObject(null);
	 * // => false
	 */
	function isObject(value) {
	  var type = typeof value;
	  return !!value && (type == 'object' || type == 'function');
	}

	/**
	 * Checks if `value` is object-like. A value is object-like if it's not `null`
	 * and has a `typeof` result of "object".
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is object-like, else `false`.
	 * @example
	 *
	 * _.isObjectLike({});
	 * // => true
	 *
	 * _.isObjectLike([1, 2, 3]);
	 * // => true
	 *
	 * _.isObjectLike(_.noop);
	 * // => false
	 *
	 * _.isObjectLike(null);
	 * // => false
	 */
	function isObjectLike(value) {
	  return !!value && typeof value == 'object';
	}

	/**
	 * Checks if `value` is classified as a `Symbol` primitive or object.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to check.
	 * @returns {boolean} Returns `true` if `value` is a symbol, else `false`.
	 * @example
	 *
	 * _.isSymbol(Symbol.iterator);
	 * // => true
	 *
	 * _.isSymbol('abc');
	 * // => false
	 */
	function isSymbol(value) {
	  return typeof value == 'symbol' ||
	    (isObjectLike(value) && objectToString.call(value) == symbolTag);
	}

	/**
	 * Converts `value` to a number.
	 *
	 * @static
	 * @memberOf _
	 * @since 4.0.0
	 * @category Lang
	 * @param {*} value The value to process.
	 * @returns {number} Returns the number.
	 * @example
	 *
	 * _.toNumber(3.2);
	 * // => 3.2
	 *
	 * _.toNumber(Number.MIN_VALUE);
	 * // => 5e-324
	 *
	 * _.toNumber(Infinity);
	 * // => Infinity
	 *
	 * _.toNumber('3.2');
	 * // => 3.2
	 */
	function toNumber(value) {
	  if (typeof value == 'number') {
	    return value;
	  }
	  if (isSymbol(value)) {
	    return NAN;
	  }
	  if (isObject(value)) {
	    var other = typeof value.valueOf == 'function' ? value.valueOf() : value;
	    value = isObject(other) ? (other + '') : other;
	  }
	  if (typeof value != 'string') {
	    return value === 0 ? value : +value;
	  }
	  value = value.replace(reTrim, '');
	  var isBinary = reIsBinary.test(value);
	  return (isBinary || reIsOctal.test(value))
	    ? freeParseInt(value.slice(2), isBinary ? 2 : 8)
	    : (reIsBadHex.test(value) ? NAN : +value);
	}

	var lodash_debounce = debounce;

	/**
	 * Returns true if value is number
	 * @param value
	 */
	function isNum(value) {
	    return typeof value === "number";
	}

	/**
	 * return true when window is defined
	 */
	function isBrowser() {
	    return typeof window !== 'undefined';
	}

	/**
	 * returns true if input is an image element
	 * @param i
	 */
	function isImage(i) {
	    return i instanceof HTMLImageElement;
	}

	/**
	 * @namespace
	 * @description Updates the src with the size of the parent element and triggers a resize event for
	 * subsequent resizing.
	 * @param steps {number | number[]} The step size in pixels.
	 * | number[] A set of image sizes in pixels.
	 * @return {Plugin}
	 * @example
	 * <AdvancedImage cldImg={img} plugins=[(responsive(100))] plugins=[(responsive([800, 1000, 1400]))] />
	 */
	function responsive(steps) {
	    return responsivePlugin.bind(null, steps);
	}
	/**
	 * @description Responsive plugin
	 * @param steps {number | number[]} The step size in pixels.
	 * | number[] A set of image sizes in pixels.
	 * @param element {HTMLImageElement} The image element
	 * @param responsiveImage {CloudinaryImage}
	 * @param htmlPluginState {HtmlPluginState} holds cleanup callbacks and event subscriptions
	 */
	function responsivePlugin(steps, element, responsiveImage, htmlPluginState) {
	    if (!isBrowser())
	        return;
	    if (!isImage(element))
	        return;
	    return new Promise(function (resolve) {
	        htmlPluginState.cleanupCallbacks.push(function () {
	            window.removeEventListener("resize", resizeRef);
	            resolve('canceled');
	        });
	        var containerSize = element.parentElement.clientWidth;
	        responsiveImage.resize(scale().width(containerSize).setActionTag('responsive'));
	        var resizeRef;
	        htmlPluginState.pluginEventSubscription.push(function () {
	            window.addEventListener('resize', resizeRef = lodash_debounce(function () {
	                onResize(steps, element, responsiveImage);
	            }, 100));
	        });
	        resolve();
	    });
	}
	/**
	 * On resize updates image src
	 * @param steps {number | number[]} The step size in pixels.
	 * | number[] A set of image sizes in pixels.
	 * @param element {HTMLImageElement} The image element
	 * @param responsiveImage {CloudinaryImage}
	 */
	function onResize(steps, element, responsiveImage) {
	    updateByContainerWidth(steps, element, responsiveImage);
	    element.src = responsiveImage.toURL();
	}
	/**
	 * Updates the responsiveImage by container width.
	 * @param steps {number | number[]} The step size in pixels.
	 * | number[] A set of image sizes in pixels.
	 * @param element {HTMLImageElement} The image element
	 * @param responsiveImage {CloudinaryImage}
	 */
	function updateByContainerWidth(steps, element, responsiveImage) {
	    var resizeValue = element.parentElement.clientWidth;
	    if (isNum(steps)) {
	        resizeValue = Math.ceil(resizeValue / steps) * steps;
	    }
	    else if (Array.isArray(steps)) {
	        resizeValue = steps.reduce(function (prev, curr) {
	            return (Math.abs(curr - resizeValue) < Math.abs(prev - resizeValue) ? curr : prev);
	        });
	    }
	    responsiveImage.transformation.actions.forEach(function (action, index) {
	        if (action instanceof Action && action.getActionTag() === 'responsive') {
	            responsiveImage.transformation.actions[index] = scale(resizeValue).setActionTag('responsive');
	        }
	    });
	}

	/**
	 * @namespace
	 * @description Loads an image once it is in a certain margin in the viewport. This includes vertical and horizontal scrolling.
	 * @param rootMargin {string} The root element's bounding box before the intersection test is performed. Default: 0px.
	 * @param threshold {number} The percentage of the image's visibility at which point the image should load. Default: 0.1 (10%).
	 * @return {Plugin}
	 * @example
	 * <AdvancedImage cldImg={img} plugins=[(lazyload('0px', 0.25))]/>
	 */
	function lazyload(rootMargin, threshold) {
	    return lazyloadPlugin.bind(null, rootMargin, threshold);
	}
	/**
	 * @description lazyload plugin
	 * @param rootMargin {string} The root element's bounding box before the intersection test is performed. Default: 0px.
	 * @param threshold {number} The percentage of the image's visibility at which point the image should load. Default: 0.1 (10%).
	 * @param element The image element.
	 * @param element {HTMLImageElement} The image element.
	 * @param cloudinaryImage {CloudinaryImage}
	 * @param htmlPluginState {HtmlPluginState} Holds cleanup callbacks and event subscriptions.
	 */
	function lazyloadPlugin(rootMargin, threshold, element, cloudinaryImage, htmlPluginState) {
	    if (rootMargin === void 0) { rootMargin = '0px'; }
	    if (threshold === void 0) { threshold = 0.1; }
	    // if SSR skip plugin
	    if (!isBrowser())
	        return false;
	    return new Promise(function (resolve) {
	        var onIntersect = function () { return (resolve()); };
	        var unobserve = detectIntersection(element, onIntersect, rootMargin, threshold);
	        htmlPluginState.cleanupCallbacks.push(function () {
	            unobserve();
	            resolve('canceled');
	        });
	    });
	}
	/**
	 * Check if IntersectionObserver is supported
	 * @return {boolean} true if window.IntersectionObserver is defined
	 */
	function isIntersectionObserverSupported() {
	    // Check that 'IntersectionObserver' property is defined on window
	    return window && 'IntersectionObserver' in window;
	}
	/**
	 * Calls onIntersect() to resolve when intersection is detected, or when
	 * no native lazy loading or when IntersectionObserver isn't supported.
	 * @param {Element} el - the element to observe
	 * @param {function} onIntersect - called when the given element is in view
	 * @param rootMargin {string} The root element's bounding box before the intersection test is performed. Default: 0px.
	 * @param threshold {number} The percentage of the image's visibility at which point the image should load. Default: 0.1 (10%).
	 */
	function detectIntersection(el, onIntersect, rootMargin, threshold) {
	    try {
	        if (!isIntersectionObserverSupported()) {
	            // Return if there's no need or possibility to detect intersection
	            onIntersect();
	            return;
	        }
	        // Detect intersection with given element using IntersectionObserver
	        var observer_1 = new IntersectionObserver(function (entries) {
	            entries.forEach(function (entry) {
	                if (entry.isIntersecting) {
	                    observer_1.unobserve(entry.target);
	                    onIntersect();
	                }
	            });
	        }, { rootMargin: rootMargin, threshold: threshold });
	        observer_1.observe(el);
	        return function () { el && observer_1.observe(el); };
	    }
	    catch (e) {
	        onIntersect();
	    }
	}

	/**
	 * @namespace
	 * @description Appends accessibility transformations to the original image.
	 * @param mode {AccessibilityMode} The accessibility mode to use. Possible modes: 'darkmode' | 'brightmode' | 'monochrome' | 'colorblind'. Default: 'darkmode'.
	 * @return {Plugin}
	 * @example
	 * <AdvancedImage cldImg={img} plugins={[accessibility()]}/>
	 */
	function accessibility(mode) {
	    if (mode === void 0) { mode = 'darkmode'; }
	    return accessibilityPlugin.bind(null, mode);
	}
	/**
	 * @description Accessibility plugin
	 * @param mode {accessibilityMode} The accessibility mode to use. Possible modes: 'darkmode' | 'brightmode' | 'monochrome' | 'colorblind'. Default: 'darkmode'.
	 * @param element {HTMLImageElement} The image element.
	 * @param pluginCloudinaryImage {CloudinaryImage}
	 * @param htmlPluginState {htmlPluginState} Holds cleanup callbacks and event subscriptions.
	 */
	function accessibilityPlugin(mode, element, pluginCloudinaryImage, htmlPluginState) {
	    if (isBrowser()) {
	        if (!isImage(element))
	            return;
	        return new Promise(function (resolve) {
	            // resolved promise when canceled
	            htmlPluginState.cleanupCallbacks.push(function () {
	                resolve('canceled');
	            });
	            if (!ACCESSIBILITY_MODES[mode]) {
	                mode = 'darkmode';
	            }
	            pluginCloudinaryImage.effect(ACCESSIBILITY_MODES[mode]);
	            resolve();
	        });
	    }
	    else {
	        pluginCloudinaryImage.effect(ACCESSIBILITY_MODES[mode]);
	    }
	}

	/**
	 * @namespace
	 * @description Displays a placeholder image until the original image loads.
	 * @param mode {PlaceholderMode} The type of placeholder image to display. Possible modes: 'vectorize' | 'pixelate' | 'blur' | 'predominant-color'. Default: 'vectorize'.
	 * @return {Plugin}
	 * @example
	 * <AdvancedImage cldImg={img} plugins=[(placeholder('blur'))]/>
	 */
	function placeholder(mode) {
	    if (mode === void 0) { mode = 'vectorize'; }
	    return placeholderPlugin.bind(null, mode);
	}
	/**
	 * @description Placeholder plugin
	 * @param mode {placeholderMode} The type of placeholder image to display. Possible modes: 'vectorize' | 'pixelate' | 'blur' | 'predominant-color'. Default: 'vectorize'.
	 * @param element {HTMLImageElement} The image element.
	 * @param pluginCloudinaryImage {CloudinaryImage}
	 * @param htmlPluginState {htmlPluginState} Holds cleanup callbacks and event subscriptions.
	 */
	function placeholderPlugin(mode, element, pluginCloudinaryImage, htmlPluginState) {
	    if (!isBrowser())
	        return;
	    if (!isImage(element))
	        return;
	    var placeholderTransformation = preparePlaceholderTransformation(mode, pluginCloudinaryImage);
	    element.src = placeholderTransformation.toURL();
	    //if placeholder does not load, load a single transparent pixel
	    element.onerror = function () {
	        element.src = singleTransparentPixel;
	    };
	    /*
	    Placeholder image loads first. Once it loads, the promise is resolved and the
	    larger image will load. Once the larger image loads, promised and plugin is resolved.
	     */
	    return new Promise(function (resolve) {
	        element.onload = function () {
	            resolve();
	        };
	    }).then(function () {
	        return new Promise(function (resolve) {
	            htmlPluginState.cleanupCallbacks.push(function () {
	                element.src = singleTransparentPixel;
	                resolve('canceled');
	            });
	            // load image once placeholder is done loading
	            var largeImage = new Image();
	            largeImage.src = pluginCloudinaryImage.toURL();
	            largeImage.onload = function () {
	                resolve();
	            };
	            // image does not load, resolve
	            largeImage.onerror = function () {
	                resolve();
	            };
	        });
	    });
	}
	/**
	 * Prepares placeholder transformation by appending a placeholder-type transformation to the end of the URL
	 * @param mode {PlaceholderMode} The type of placeholder image to display. Possible modes: 'vectorize' | 'pixelate' | 'blur' | 'predominant-color'. Default: 'vectorize'.
	 * @param pluginCloudinaryImage {CloudinaryImage}
	 */
	function preparePlaceholderTransformation(mode, pluginCloudinaryImage) {
	    var placeholderClonedImage = cloneDeep_1(pluginCloudinaryImage);
	    // @ts-ignore
	    if (!PLACEHOLDER_IMAGE_OPTIONS[mode]) {
	        mode = 'vectorize';
	    }
	    //appends a placeholder transformation on placeholderClonedImage
	    // @ts-ignore
	    PLACEHOLDER_IMAGE_OPTIONS[mode].actions.forEach(function (transformation) {
	        placeholderClonedImage.addAction(transformation);
	    });
	    return placeholderClonedImage;
	}

	function serverSideSrc(plugins, serverCloudinaryImage) {
	    var clonedServerCloudinaryImage = cloneDeep_1(serverCloudinaryImage);
	    if (plugins) {
	        for (var i = 0; i < plugins.length; i++) {
	            var response = plugins[i](null, clonedServerCloudinaryImage);
	            if (!response) { //lazyload
	                break;
	            }
	        }
	    }
	    return clonedServerCloudinaryImage.toURL();
	}

	/**
	 * Cancels currently running plugins. This is called from unmount or update
	 * @param pluginState {HtmlPluginState} Holds cleanup callbacks and event subscriptions
	 */
	function cancelCurrentlyRunningPlugins(pluginState) {
	    pluginState.cleanupCallbacks.forEach(function (fn) {
	        fn(); // resolve each promise with 'canceled'
	    });
	}

	exports.HtmlImageLayer = HtmlImageLayer;
	exports.HtmlPictureLayer = HtmlPictureLayer;
	exports.HtmlVideoLayer = HtmlVideoLayer;
	exports.accessibility = accessibility;
	exports.cancelCurrentlyRunningPlugins = cancelCurrentlyRunningPlugins;
	exports.isBrowser = isBrowser;
	exports.lazyload = lazyload;
	exports.placeholder = placeholder;
	exports.responsive = responsive;
	exports.serverSideSrc = serverSideSrc;

	Object.defineProperty(exports, '__esModule', { value: true });

})));
